---
title: Quality assessment of preeclampsia data
author:
- name: Berta Canal SimÃ³n
  affiliation:
  - &id Barcelona School of Economics
  email: berta.canal@bse.eu
- name: Adam Olivares Canal
  affiliation: *id
  email: adam.olivares@bse.eu
date: "`r format(Sys.time(), '%B %e, %Y')`"
abstract: >
  Here we perform a first exploratory analysis and quality assessment of the cfRNA expression profiles of the preeclampsia data.
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    fig_captions: yes
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: _bibliography.bib
---

```{r setup, echo=FALSE, cache=FALSE}
library(knitr)
library(kableExtra)


knitr::opts_chunk$set(
  collapse=TRUE,
  comment="",
  fig.align="center",
  fig.wide=TRUE,
  cache=TRUE,
  cache.path="_cache/__QA_pe",
  cache.extra=R.version.string,
  autodep=TRUE
)

source(file.path("_scripts", "labelPanel.R")) ## for the labelPanel() function
source(file.path("_scripts", "myplotbcv.R")) ## for the myplotbcv() function
```

# Read molecular and phenotype data

Read the molecular and phenotype data stored in serialized `SummarizedExperiment`
objects.

```{r, message=FALSE}
library(SummarizedExperiment)

seM <- readRDS(file.path("_raw_data", "se.Moufarrej22.GENCODEv44.hg38.rds"))
dim(seM)
seD <- readRDS(file.path("_raw_data", "se.DelVecchio21.GENCODEv44.hg38.rds"))
dim(seD)
```

# Filtering of lowly-expressed genes and normalization

We will use the [edgeR](https://bioconductor.org/packages/edgeR) package for
processing the raw counts and doing some exploratory analyses. We start by
building the corresponding `DGEList` objects with the corresponding row and
column data.

```{r, message=FALSE, warning=FALSE}
library(edgeR)

dgeM <- DGEList(counts=assay(seM), genes=rowData(seM), samples=colData(seM))
dgeD <- DGEList(counts=assay(seD), genes=rowData(seD), samples=colData(seD))
```

Filter lowly expressed genes using the function `filterByExpr()` and the
`Preeclampsia` phenotype column as grouping factor. We need to filter both,
the `SummarizedExperiment` and `DGEList` types of objects, to keep them
parallel to each other.

```{r}
mask <- filterByExpr(dgeM, group=seM$Preeclampsia)
sum(mask)
seM.filt <- seM[mask, ]
dim(seM.filt)
dgeM.filt <- dgeM[mask, , keep.lib.sizes=FALSE]
dim(dgeM.filt)

mask <- filterByExpr(dgeD, group=seD$Preeclampsia)
sum(mask)
seD.filt <- seD[mask, ]
dim(seD.filt)
dgeD.filt <- dgeD[mask, , keep.lib.sizes=FALSE]
dim(dgeD.filt)
```

Calculate normalization factors on the filtered data.

```{r}
dgeM.filt <- calcNormFactors(dgeM.filt)
assays(seM.filt)$logCPM <- cpm(dgeM.filt, log=TRUE)

dgeD.filt <- calcNormFactors(dgeD.filt)
assays(seD.filt)$logCPM <- cpm(dgeD.filt, log=TRUE)
```

Divide samples in the [@moufarrej2022early] study in two subsets, one for
_training_ in which we will put together the samples from the `Discovery` and
`Validation1` cohorts, and another one for _testing_ formed by the samples of
the `Validation2` cohort. In both cases, we exclude samples drawn after
delivery, labeled as `Post-partum` in the `SamplingGAgroup` column.

```{r}
table(seM.filt$Cohort)
mask <- seM.filt$Cohort %in% c("Discovery", "Validation1") &
        !seM.filt$SamplingGAgroup %in% "Post-partum"
seM.filt.training <- seM.filt[, mask]
dim(seM.filt.training)
dgeM.filt.training <- dgeM.filt[, mask]
dim(dgeM.filt.training)
mask <- seM.filt$Cohort %in% "Validation2" &
        !seM.filt$SamplingGAgroup %in% "Post-partum"
seM.filt.testing <- seM.filt[, mask]
dim(seM.filt.testing)
dgeM.filt.testing <- dgeM.filt[, mask]
dim(dgeM.filt.testing)
```

From the study by [@delvecchio2021cell], discard samples from non-pregnant
women, drawn after delivery, or taken from the cord blood of the neonate.

```{r}
table(seD.filt$SamplingGAgroup)
mask <- !grepl("Non pregnant", seD.filt$SamplingGAgroup) &
        !seD.filt$SamplingGAgroup %in% c("delivery", "cord blood")
seD.filt.subset <- seD.filt[, mask]
dim(seD.filt.subset)
dgeD.filt.subset <- dgeD.filt[, mask]
dim(dgeD.filt.subset)
```

Figure \@ref(fig:mdsplot) below shows samples dissimilarity labeled by
the diagnosis of preeclampsia in each of the three datasets.

```{r mdsplot, echo=FALSE, fig.height=4, fig.width=12, out.width="800px", fig.cap="Sample dissimilarity labeled by the diagnosis of preeclampsia. Multidimensional scaling (MDS) plot of sample dissimilarity for the training (a) and testing (b) data from the study by @moufarrej2022early, and for the data from the study by [@delvecchio2021cell]."}
preeclampsiacolors <- c("skyblue", "darkred")
par(mfrow=c(1, 3), mar=c(4, 5, 2, 1))
xx <- plotMDS(dgeM.filt.training, labels=rep("", ncol(seM.filt.training)), las=1)
points(xx$x, xx$y, col=preeclampsiacolors[as.integer(seM.filt.training$Preeclampsia)],
       pch=19, cex=0.8)
legend("topleft", c("Control", "Preeclampsia"), fill=preeclampsiacolors,
       inset=0.01)
labelPanel("a")
xx <- plotMDS(dgeM.filt.testing, labels=rep("", ncol(seM.filt.testing)), las=1)
points(xx$x, xx$y, col=preeclampsiacolors[as.integer(seM.filt.testing$Preeclampsia)],
       pch=19, cex=0.8)
legend("topleft", c("Control", "Preeclampsia"), fill=preeclampsiacolors,
       inset=0.01)
labelPanel("b")
xx <- plotMDS(dgeD.filt.subset, labels=rep("", ncol(seD.filt.subset)), las=1)
points(xx$x, xx$y, col=preeclampsiacolors[as.integer(seD.filt.subset$Preeclampsia)],
       pch=19, cex=0.8)
legend("topright", c("Control", "Preeclampsia"), fill=preeclampsiacolors,
       inset=0.01)
labelPanel("c")
```

Save `DGEList` and `SummarizedExperiment` objects including filtered data,
normalization factors and normalized log CPM units of expression.

```{r}
saveRDS(dgeM.filt, file=file.path("_processed_data", "dgeM.filt.rds"))
saveRDS(seM.filt, file=file.path("_processed_data", "seM.filt.rds"))

saveRDS(dgeD.filt, file=file.path("_processed_data", "dgeD.filt.rds"))
saveRDS(seD.filt, file=file.path("_processed_data", "seD.filt.rds"))

saveRDS(dgeM.filt.training,
        file=file.path("_processed_data", "dgeM.filt.training.rds"))
saveRDS(seM.filt.training,
        file=file.path("_processed_data", "seM.filt.training.rds"))
saveRDS(dgeM.filt.testing,
        file=file.path("_processed_data", "dgeM.filt.testing.rds"))
saveRDS(seM.filt.testing,
        file=file.path("_processed_data", "seM.filt.testing.rds"))

saveRDS(dgeD.filt.subset,
        file=file.path("_processed_data", "dgeD.filt.subset.rds"))
saveRDS(seD.filt.subset,
        file=file.path("_processed_data", "seD.filt.subset.rds"))
```

# Dispersion estimates and hypervariable genes

We calculate dispersion estimates and biological coefficients of variation
[@mccarthy2012differential] and use them to identify hypervariable genes for
different models of increasing complexity that will help us to select the most
sensible one for the differential expression analysis between
preeclampsia-affected and unafected mothers.

The four considered models are: (1) a model with the intercept term only,
(2) a model with `Preeclampsia` as main explanatory variable, (3) a model
with `Preeclampsia` as main explanatory variable and `SamplingGA` as a
covariate, and (4) a model with `Preeclampsia` as main explanatory variable,
and `SamplingGA` and surrogate variables (SVs), calculated with SVA
[@leek2007capturing] as covariates. The estimation of SVs will be done using
only the expression values of the top-10% most variable genes.

## Training dataset from Moufarrej et al. (2022)

```{r, message=FALSE}
library(sva)

mod <- model.matrix(~ 1, colData(seM.filt.training))
dgeM.filt.training.interceptonly <- estimateDisp(dgeM.filt.training, design=mod,
                                                 robust=TRUE)
summary(dgeM.filt.training.interceptonly$prior.df)
mod <- model.matrix(~ Preeclampsia, colData(seM.filt.training))
dgeM.filt.training.PE <- estimateDisp(dgeM.filt.training, design=mod,
                                      robust=TRUE)
summary(dgeM.filt.training.PE$prior.df)
mod <- model.matrix(~ Preeclampsia + SamplingGA, colData(seM.filt.training))
dgeM.filt.training.PEGA <- estimateDisp(dgeM.filt.training, design=mod,
                                        robust=TRUE)
summary(dgeM.filt.training.PEGA$prior.df)
mod0 <- model.matrix(~ SamplingGA, colData(seM.filt.training))
IQRs <- apply(assays(seM.filt.training)$logCPM, 1, IQR)
mask <- IQRs > quantile(IQRs, prob=0.9)
sv <- sva(assays(seM.filt.training)$logCPM[mask, ],
          mod=mod, mod0=mod0)
mod <- cbind(mod, sv$sv)
dgeM.filt.training.PEGASVs <- estimateDisp(dgeM.filt.training,
                                           design=mod, robust=TRUE)
summary(dgeM.filt.training.PEGASVs$prior.df)
```

Figure \@ref(fig:Mtrainingavglogcpmxvbcv) shows, for each considered model, the
biological coefficient of variation (BCV) as function of the average expression,
in log CPM units.

```{r Mtrainingavglogcpmxvbcv, echo=FALSE, fig.height=8, fig.width=8, dpi=100, results="hide", fig.cap="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units."}
rngbcv <- range(sqrt(c(getDispersion(dgeM.filt.training.interceptonly),
                       getDispersion(dgeM.filt.training.PE),
                       getDispersion(dgeM.filt.training.PEGA),
                       getDispersion(dgeM.filt.training.PEGASVs))))
rngbcv[2] <- ceiling(rngbcv[2])
par(mfrow=c(2, 2), mar=c(4, 5, 2, 1))
syms <- mcols(seM.filt.training)$Symbol
myplotbcv(dgeM.filt.training.interceptonly, "~ 1", rngbcv, syms, cutoffdisp1=1.5,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeM.filt.training.PE, "~ Preeclampsia", rngbcv, syms, cutoffdisp1=1.4,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeM.filt.training.PEGA, "~ Preeclampsia + SamplingGA", rngbcv, syms,
          cutoffdisp1=1.4, cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeM.filt.training.PEGASVs, "~ Preeclampsia + SamplingGA + SVs",
          rngbcv, syms, cutoffdisp1=0.9, cutoffdisp2=0.8, cutoffA=9)
```

Table \@ref(tab:hypervargenesMtraining) below shows the top-20 hypervariable genes
with lowest prior degrees of freedom for the last model.

```{r hypervargenesMtraining, echo=FALSE}
ord <- order(dgeM.filt.training.PEGASVs$prior.df)
dtf <- data.frame(prior.df=dgeM.filt.training.PEGASVs$prior.df[ord],
                  BCV=sqrt(dgeM.filt.training.PEGASVs$tagwise.dispersion[ord]),
                  Gene=dgeM.filt.training.PEGASVs$genes$Symbol[ord],
                  Description=dgeM.filt.training.PEGASVs$genes$Description[ord])
ktab <- kable(dtf[1:20, ], "html", escape=FALSE, row.names=FALSE,
              caption="Top-20 hypervariable genes for the model Preeclampsia + SamplingGA + SVs.")
ktab <- kable_styling(ktab, bootstrap_options=c("striped", "hover", "responsive"),
                      fixed_thead=TRUE)
kable_styling(ktab, position="center")
```

## Testing dataset from Moufarrej et al. (2022)

```{r, message=FALSE}
mod <- model.matrix(~ 1, colData(seM.filt.testing))
dgeM.filt.testing.interceptonly <- estimateDisp(dgeM.filt.testing, design=mod,
                                                robust=TRUE)
summary(dgeM.filt.testing.interceptonly$prior.df)
mod <- model.matrix(~ Preeclampsia, colData(seM.filt.testing))
dgeM.filt.testing.PE <- estimateDisp(dgeM.filt.testing, design=mod,
                                     robust=TRUE)
summary(dgeM.filt.testing.PE$prior.df)
mod <- model.matrix(~ Preeclampsia + SamplingGA, colData(seM.filt.testing))
dgeM.filt.testing.PEGA <- estimateDisp(dgeM.filt.testing, design=mod,
                                       robust=TRUE)
summary(dgeM.filt.testing.PEGA$prior.df)
mod0 <- model.matrix(~ SamplingGA, colData(seM.filt.testing))
IQRs <- apply(assays(seM.filt.testing)$logCPM, 1, IQR)
mask <- IQRs > quantile(IQRs, prob=0.9)
sv <- sva(assays(seM.filt.testing)$logCPM[mask, ],
          mod=mod, mod0=mod0)
mod <- cbind(mod, sv$sv)
dgeM.filt.testing.PEGASVs <- estimateDisp(dgeM.filt.testing,
                                          design=mod, robust=TRUE)
summary(dgeM.filt.testing.PEGASVs$prior.df)
```

Figure \@ref(fig:Mtestingavglogcpmxvbcv) shows, for each considered model, the
biological coefficient of variation (BCV) as function of the average expression,
n log CPM units.

```{r Mtestingavglogcpmxvbcv, echo=FALSE, fig.height=8, fig.width=8, dpi=100, results="hide", fig.cap="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units."}
rngbcv <- range(sqrt(c(getDispersion(dgeM.filt.testing.interceptonly),
                       getDispersion(dgeM.filt.testing.PE),
                       getDispersion(dgeM.filt.testing.PEGA),
                       getDispersion(dgeM.filt.testing.PEGASVs))))
rngbcv[2] <- ceiling(rngbcv[2])
par(mfrow=c(2, 2), mar=c(4, 5, 2, 1))
syms <- mcols(seM.filt.testing)$Symbol
myplotbcv(dgeM.filt.testing.interceptonly, "~ 1", rngbcv, syms, cutoffdisp1=1.4,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeM.filt.testing.PE, "~ Preeclampsia", rngbcv, syms, cutoffdisp1=1.4,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeM.filt.testing.PEGA, "~ Preeclampsia + SamplingGA", rngbcv, syms,
          cutoffdisp1=1.4, cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeM.filt.testing.PEGASVs, "~ Preeclampsia + SamplingGA + SVs",
          rngbcv, syms, cutoffdisp1=1, cutoffdisp2=0.6, cutoffA=8)
```

## Subset from Del Vecchio et al. (2021)

```{r, message=FALSE}
mod <- model.matrix(~ 1, colData(seD.filt.subset))
dgeD.filt.subset.interceptonly <- estimateDisp(dgeD.filt.subset, design=mod,
                                               robust=TRUE)
summary(dgeD.filt.subset.interceptonly$prior.df)
mod <- model.matrix(~ Preeclampsia, colData(seD.filt.subset))
dgeD.filt.subset.PE <- estimateDisp(dgeD.filt.subset, design=mod,
                                    robust=TRUE)
summary(dgeD.filt.subset.PE$prior.df)
mod <- model.matrix(~ Preeclampsia + SamplingGA, colData(seD.filt.subset))
dgeD.filt.subset.PEGA <- estimateDisp(dgeD.filt.subset, design=mod,
                                      robust=TRUE)
summary(dgeD.filt.subset.PEGA$prior.df)
mod0 <- model.matrix(~ SamplingGA, colData(seD.filt.subset))
IQRs <- apply(assays(seD.filt.subset)$logCPM, 1, IQR)
mask <- IQRs > quantile(IQRs, prob=0.9)
sv <- sva(assays(seD.filt.subset)$logCPM[mask, ],
          mod=mod, mod0=mod0)
mod <- cbind(mod, sv$sv)
dgeD.filt.subset.PEGASVs <- estimateDisp(dgeD.filt.subset,
                                         design=mod, robust=TRUE)
summary(dgeD.filt.subset.PEGASVs$prior.df)
```

Figure \@ref(fig:Dsubsetavglogcpmxvbcv) shows, for each considered model, the
biological coefficient of variation (BCV) as function of the average expression,
n log CPM units.

```{r Dsubsetavglogcpmxvbcv, echo=FALSE, fig.height=8, fig.width=8, dpi=100, results="hide", fig.cap="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units."}
rngbcv <- range(sqrt(c(getDispersion(dgeD.filt.subset.interceptonly),
                       getDispersion(dgeD.filt.subset.PE),
                       getDispersion(dgeD.filt.subset.PEGA),
                       getDispersion(dgeD.filt.subset.PEGASVs))))
rngbcv[2] <- ceiling(rngbcv[2])
par(mfrow=c(2, 2), mar=c(4, 5, 2, 1))
syms <- mcols(seD.filt.subset)$Symbol
myplotbcv(dgeD.filt.subset.interceptonly, "~ 1", rngbcv, syms, cutoffdisp1=1.4,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeD.filt.subset.PE, "~ Preeclampsia", rngbcv, syms, cutoffdisp1=1.4,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeD.filt.subset.PEGA, "~ Preeclampsia + SamplingGA", rngbcv, syms,
          cutoffdisp1=1.4, cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeD.filt.subset.PEGASVs, "~ Preeclampsia + SamplingGA + SVs",
          rngbcv, syms, cutoffdisp1=1, cutoffdisp2=0.6, cutoffA=8)
```

Table \@ref(tab:hypervargenesDsubset) below shows the top-20 hypervariable genes
with lowest prior degrees of freedom for the last model.

```{r hypervargenesDsubset, echo=FALSE}
ord <- order(dgeD.filt.subset.PEGASVs$prior.df)
dtf <- data.frame(prior.df=dgeD.filt.subset.PEGASVs$prior.df[ord],
                  BCV=sqrt(dgeD.filt.subset.PEGASVs$tagwise.dispersion[ord]),
                  Gene=dgeD.filt.subset.PEGASVs$genes$Symbol[ord],
                  Description=dgeD.filt.subset.PEGASVs$genes$Description[ord])
ktab <- kable(dtf[1:20, ], "html", escape=FALSE, row.names=FALSE,
              caption="Top-20 hypervariable genes for the model Preeclampsia + SamplingGA + SVs.")
ktab <- kable_styling(ktab, bootstrap_options=c("striped", "hover", "responsive"),
                      fixed_thead=TRUE)
kable_styling(ktab, position="center")
```
# Session information

```{r, child="_session-info.Rmd"}
```

# References

