---
title: Quality assessment of cancer data
author:
- name: Berta Canal Simón
  affiliation:
  - &id Barcelona School of Economics
  email: berta.canal@bse.eu
- name: Adam Olivares Canal
  affiliation: *id
  email: adam.olivares@bse.eu
date: "`r format(Sys.time(), '%B %e, %Y')`"
abstract: >
  Here we perform a first exploratory analysis and quality assessment of the cfRNA expression profiles of the cancer data.
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    fig_captions: yes
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: _bibliography.bib
---

```{r setup, echo=FALSE, cache=FALSE, message=FALSE}
library(knitr)
library(kableExtra)
library(VIM)

knitr::opts_chunk$set(
  collapse=TRUE,
  comment="",
  fig.align="center",
  fig.wide=TRUE,
  cache=TRUE,
  cache.path="_cache/__QA_cn",
  cache.extra=R.version.string,
  autodep=TRUE
)

source(file.path("_scripts", "labelPanel.R")) ## for the labelPanel() function
source(file.path("_scripts", "myplotbcv.R")) ## for the myplotbcv() function
```

# Read molecular and phenotype data

Read the molecular and phenotype data stored in serialized `SummarizedExperiment`
objects.

```{r, message=FALSE}
library(SummarizedExperiment)

seR <- readRDS(file.path("_raw_data", "se.Roskams22.GENCODEv44.hg38.rds"))
dim(seR)
seB <- readRDS(file.path("_raw_data", "se.Block22.GENCODEv44.hg38.rds"))
dim(seB)
```


# Filtering of lowly-expressed genes and normalization

In [@Roskams2022], they sequenced plasma samples of eight hepatocellular carcinoma (HCC) and ten multiple myeloma (MM) patients, 12 patients of their respective precancerous conditions, and 20 non-cancer (NC) donors. To avoid within group heterogeneity, our study focuses on individuals with hepatocellular carcinoma (HCC) and non-cancer (NC) for the control. Otherwise, our analysis could increase between group variance and thus lead to spurious results.

NOTA: See abstract from the paper and suplementary tables to modify this paragraph.

```{r}
table(colData(seR)$Disease)
mask <- colData(seR)$Disease %in% c('NC', 'HCC')
seR.subset <- seR[,mask]
dim(seR.subset)
colData(seR.subset)$Tumor <- factor(seR.subset$Disease == 'HCC', labels = c('no', 'yes'))
table(seR.subset$Disease)
```
For [@Block2022], the project also focus on individuals with hepatocellular carcinoma (HCC) in plasma and plasma related EVs and non-cancer (NC) for the control.

```{r}
table(seB$SampleType)
table(seB$SampleType, seB$Tissue)
mask <- (seB$SampleType == "HCC" &
         (seB$Tissue %in% c("plasma", "plasma_vesicles"))) |
        seB$SampleType == "NHC"
seB.subset <- seB[,mask]
dim(seB.subset)
colData(seB.subset)$Tumor <- factor(seB.subset$SampleType == 'HCC',
                                    labels = c('no', 'yes'))
table(seB.subset$SampleType)
```

We will use the [edgeR](https://bioconductor.org/packages/edgeR) package for
processing the raw counts and doing some exploratory analyses. We start by
building the corresponding `DGEList` objects with the corresponding row and
column data.

```{r, message = FALSE, warning = FALSE}
library(edgeR)

dgeR <- DGEList(counts=assay(seR.subset), genes=rowData(seR.subset),
                samples=colData(seR.subset))
dgeB <- DGEList(counts=assay(seB.subset), genes=rowData(seB.subset),
                samples=colData(seB.subset))
```

Filter lowly expressed genes using the function `filterByExpr()` and the
`Tumor` phenotype column as grouping factor. We need to filter both,
the `SummarizedExperiment` and `DGEList` types of objects, to keep them
parallel to each other.

```{r}
mask <- filterByExpr(dgeR, group=seR.subset$Tumor)
sum(mask)
seR.filt <- seR.subset[mask, ]
dim(seR.filt)
dgeR.filt <- dgeR[mask, , keep.lib.sizes=FALSE]
dim(dgeR.filt)

mask <- filterByExpr(dgeB, group=seB.subset$Tumor)
sum(mask)
seB.filt <- seB.subset[mask, ]
dim(seB.filt)
dgeB.filt <- dgeB[mask, , keep.lib.sizes=FALSE]
dim(dgeB.filt)
```

Calculate normalization factors on the filtered data.

```{r}
dgeR.filt <- calcNormFactors(dgeR.filt)
assays(seR.filt, withDimnames=FALSE)$logCPM <- cpm(dgeR.filt, log=TRUE)

dgeB.filt <- calcNormFactors(dgeB.filt)
assays(seB.filt)$logCPM <- cpm(dgeB.filt, log=TRUE)
```

Fo simplicity, change the name of the data from articles [@Roskams2022] and [@Block2022] to _training_ and _testing_, respectively.
```{r}
# Si no treballem amb aquests noms millor que els treiguem, però si els treiem hem de modificar els objectes guardats en aquest html i els carregats en el següent.
seR.filt.training <- seR.filt
dgeR.filt.training <- dgeR.filt

seB.filt.testing <- seB.filt
dgeB.filt.testing <- dgeB.filt
```

Figure \@ref(fig:mdsplot) below shows samples dissimilarity labeled by
the diagnosis of cancer in each of the two datasets.

```{r mdsplot, echo=FALSE, fig.height=4, fig.width=12, out.width="800px", fig.cap="Sample dissimilarity labeled by the diagnosis of cancer. Multidimensional scaling (MDS) plot of sample dissimilarity for the training (a) and testing (b) data from the study by @Roskams2022, and for the data from the study by @Block2022."}
tumorcolors <- c("skyblue", "darkred")
par(mfrow=c(1, 2), mar=c(4, 5, 2, 1))
xx <- plotMDS(dgeR.filt.training, labels=rep("", ncol(seR.filt.training)), las=1)
points(xx$x, xx$y, col=tumorcolors[as.integer(seR.filt.training$Tumor)],
       pch=19, cex=0.8)
legend("topleft", c("Control", " Cancer"), fill=tumorcolors,
       inset=0.01)
labelPanel("a")
xx <- plotMDS(dgeB.filt.testing, labels=rep("", ncol(seB.filt.testing)), las=1)
points(xx$x, xx$y, col=tumorcolors[as.integer(seB.filt.testing$Tumor)],
       pch=19, cex=0.8)
legend("topleft", c("Control", "Cancer"), fill=tumorcolors,
       inset=0.01)
labelPanel("b")
```



# Data imputation

## Testing dataset

### Variable 'Sex'

The estimation of missing data for `Sex` covariate uses a KNN imputer (k = 3) based on the relationship between the expression of the MSY genes (genes from the male-specific region of the Y chromosome) and XIE genes (genes that escape X-chromosome inactivation).

```{r}
library(SummarizedExperiment)
library(edgeR)

## llegim cataleg de genes amb expressio especifica de sexe documentada
MSY_genes <- readRDS("MSYgenes.rds")
XIE_genes <- readRDS("XIEgenes.rds")

## filterem aquells que no surten a les nostres dades
MSY_genes <- MSY_genes[MSY_genes %in% rownames(seB.filt.testing)]
XIE_genes <- XIE_genes[XIE_genes %in% rownames(seB.filt.testing)]

## una possible aproximacio al problema seria calcular la mitjana
## d'expressio a cada mostra separadament pels gens MSY i XIE
msyexp <- colMeans(assays(seB.filt.testing)$logCPM[MSY_genes, ])
xieexp <- colMeans(assays(seB.filt.testing)$logCPM[XIE_genes, ])
```

```{r, echo = FALSE}
## visualitzem per cada mostra la seva expressio en MSY vs XIE
## posant M per mostres "male", F per "female" i O per "missing"
rng <- range(c(msyexp, xieexp))
plot(msyexp, xieexp, pch="", xlim=rng, ylim=rng, las=1)
lab <- c("F", "M")[as.integer(factor(seB.filt.testing$Sex))]
lab[is.na(lab)] <- "O"
points(msyexp, xieexp, pch=lab)
abline(0, 1)
```

From Robert: Since cancer can cause chromosomal rearrangements, it could be that the signal we expect to see in the samples analyzed is distorted, and it should also be taken into account that these molecules of extracellular RNA are found in plasma and so they can come from a heterogeneous set of tissues that can give rise to more variability.

In any case, there are two clusters, one of "male" and the other "female" samples. In the "female" cluster there are two samples labeled as males that can be great candidates for having sex reported incorrectly, and it is quite clear what gender the "missing" could correspond to.

Correct the labeled from the two potential incorrectly classified samples.

```{r}
mask <- colnames(seB.filt.testing) %in% names(sort(msyexp[seB.filt.testing$Sex == 'male'])[1:2])
seB.filt.testing$Sex[mask] <- 'female'
```

Impute observations with missing value for sex.

```{r}
data.imputation <- as.data.frame(t(rbind(msyexp, xieexp)))
data.imputation$Sex <- as.integer(factor(seB.filt.testing$Sex))
data.imputation <- kNN(data.imputation, k=3, variable="Sex")
data.imputation$Sex <- factor(data.imputation$Sex)
levels(data.imputation$Sex) <- c("female", "male")
seB.filt.testing$Sex <- data.imputation$Sex
dgeB.filt.testing$samples$Sex <- data.imputation$Sex
```

```{r, echo = FALSE}
rng <- range(c(msyexp, xieexp))
plot(msyexp, xieexp, pch="", xlim=rng, ylim=rng, las=1)
lab <- c("F", "M")[as.integer(factor(seB.filt.testing$Sex))] # no needed to convert into an integer factor because it already is after imputation
lab[is.na(lab)] <- "O"
points(msyexp, xieexp, pch=lab)
abline(0, 1)
```

Save `DGEList` and `SummarizedExperiment` objects including filtered data,
normalization factors and normalized log CPM units of expression.

```{r}
saveRDS(dgeR.filt, file=file.path("_processed_data", "dgeR.filt.rds"))
saveRDS(seR.filt, file=file.path("_processed_data", "seR.filt.rds"))

saveRDS(dgeB.filt, file=file.path("_processed_data", "dgeB.filt.rds"))
saveRDS(seB.filt, file=file.path("_processed_data", "seB.filt.rds"))

saveRDS(dgeR.filt.training,
        file=file.path("_processed_data", "dgeR.filt.training.rds"))
saveRDS(seR.filt.training,
        file=file.path("_processed_data", "seR.filt.training.rds"))
saveRDS(dgeB.filt.testing,
        file=file.path("_processed_data", "dgeB.filt.testing.rds"))
saveRDS(seB.filt.testing,
        file=file.path("_processed_data", "seB.filt.testing.rds"))
```

# Dispersion estimates and hypervariable genes

We calculate dispersion estimates and biological coefficients of variation
[@mccarthy2012differential] and use them to identify hypervariable genes for
different models of increasing complexity that will help us to select the most
sensible one for the differential expression analysis between
Cancer-affected and unaffected individuals.

The four considered models are: (1) a model with the intercept term only,
(2) a model with `Tumor` as main explanatory variable, (3) a model
with `Tumor` as main explanatory variable and `Sex` as
covariate, and (4) a model with `Tumor` as main explanatory variable, and `Sex` and surrogate variables (SVs), calculated with SVA
[@leek2007capturing] as covariates. The estimation of SVs will be done using
only the expression values of the top-10% most variable genes.

## Training dataset

```{r, message=FALSE}
library(sva)

mod <- model.matrix(~ 1, colData(seR.filt.training))
dgeR.filt.training.interceptonly <- estimateDisp(dgeR.filt.training, design=mod,
                                                 robust=TRUE)
summary(dgeR.filt.training.interceptonly$prior.df)

mod <- model.matrix(~ Tumor, colData(seR.filt.training))
dgeR.filt.training.PE <- estimateDisp(dgeR.filt.training, design=mod,
                                      robust=TRUE)
summary(dgeR.filt.training.PE$prior.df)

mod <- model.matrix(~ Tumor + Sex, colData(seR.filt.training))
dgeR.filt.training.PEGA <- estimateDisp(dgeR.filt.training, design=mod,
                                        robust=TRUE)
summary(dgeR.filt.training.PEGA$prior.df)

mod0 <- model.matrix(~ Sex, colData(seR.filt.training))
IQRs <- apply(assays(seR.filt.training)$logCPM, 1, IQR)
mask <- IQRs > quantile(IQRs, prob=0.9)
sv <- sva(assays(seR.filt.training)$logCPM[mask, ],
          mod=mod, mod0=mod0)
mod <- cbind(mod, sv$sv)
dgeR.filt.training.PEGASVs <- estimateDisp(dgeR.filt.training,
                                           design=mod, robust=TRUE)
summary(dgeR.filt.training.PEGASVs$prior.df)
```

Figure \@ref(fig:Mtrainingavglogcpmxvbcv) shows, for each considered model, the
biological coefficient of variation (BCV) as function of the average expression,
in log CPM units.

```{r Mtrainingavglogcpmxvbcv, echo=FALSE, message = FALSE, fig.height=8, fig.width=8, dpi=100, results="hide", fig.cap="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units."}
rngbcv <- range(sqrt(c(getDispersion(dgeR.filt.training.interceptonly),
                       getDispersion(dgeR.filt.training.PE),
                       getDispersion(dgeR.filt.training.PEGA),
                       getDispersion(dgeR.filt.training.PEGASVs))))
rngbcv[2] <- ceiling(rngbcv[2])
par(mfrow=c(2, 2), mar=c(4, 5, 2, 1))
syms <- mcols(seR.filt.training)$Symbol
myplotbcv(dgeR.filt.training.interceptonly, "~ 1", rngbcv, syms, cutoffdisp1=1.5,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeR.filt.training.PE, "~ Tumor", rngbcv, syms, cutoffdisp1=1.4,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeR.filt.training.PEGA, "~ Tumor + Sex", rngbcv, syms,
          cutoffdisp1=1.4, cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeR.filt.training.PEGASVs, "~ Tumor + Sex + SVs",
          rngbcv, syms, cutoffdisp1=0.9, cutoffdisp2=0.8, cutoffA=9)
```

Table \@ref(tab:hypervargenesMtrainingcn) below shows the top-20 hypervariable genes
with lowest prior degrees of freedom for the last model.

```{r hypervargenesMtrainingcn, echo=FALSE}
ord <- order(dgeR.filt.training.PEGASVs$prior.df)
dtf <- data.frame(prior.df=dgeR.filt.training.PEGASVs$prior.df[ord],
                  BCV=sqrt(dgeR.filt.training.PEGASVs$tagwise.dispersion[ord]),
                  Gene=dgeR.filt.training.PEGASVs$genes$Symbol[ord],
                  Description=dgeR.filt.training.PEGASVs$genes$Description[ord])
ktab <- kable(dtf[1:20, ], "html", escape=FALSE, row.names=FALSE,
              caption="Top-20 hypervariable genes for the model Tumor + Sex + SVs.")
ktab <- kable_styling(ktab, bootstrap_options=c("striped", "hover", "responsive"),
                      fixed_thead=TRUE)
kable_styling(ktab, position="center")
```

## Testing data

```{r, message=FALSE}
mod <- model.matrix(~ 1, colData(seB.filt.testing))
dgeB.filt.testing.interceptonly <- estimateDisp(dgeB.filt.testing, design=mod,
                                                robust=TRUE)
summary(dgeB.filt.testing.interceptonly$prior.df)
mod <- model.matrix(~ Tumor, colData(seB.filt.testing))
dgeB.filt.testing.PE <- estimateDisp(dgeB.filt.testing, design=mod,
                                     robust=TRUE)
summary(dgeB.filt.testing.PE$prior.df)
mod <- model.matrix(~ Tumor + Sex, colData(seB.filt.testing))
dgeB.filt.testing.PEGA <- estimateDisp(dgeB.filt.testing, design=mod,
                                       robust=TRUE)
summary(dgeB.filt.testing.PEGA$prior.df)
mod0 <- model.matrix(~ Sex, colData(seB.filt.testing))
IQRs <- apply(assays(seB.filt.testing)$logCPM, 1, IQR)
mask <- IQRs > quantile(IQRs, prob=0.9)
sv <- sva(assays(seB.filt.testing)$logCPM[mask, ],
          mod=mod, mod0=mod0)
mod <- cbind(mod, sv$sv)
dgeB.filt.testing.PEGASVs <- estimateDisp(dgeB.filt.testing,
                                          design=mod, robust=TRUE)
summary(dgeB.filt.testing.PEGASVs$prior.df)
```

```{r Mtestingavglogcpmxvbcv, echo=FALSE, fig.height=8, fig.width=8, dpi=100, results="hide", fig.cap="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units."}
rngbcv <- range(sqrt(c(getDispersion(dgeB.filt.testing.interceptonly),
                       getDispersion(dgeB.filt.testing.PE),
                       getDispersion(dgeB.filt.testing.PEGA),
                       getDispersion(dgeB.filt.testing.PEGASVs))))
rngbcv[2] <- ceiling(rngbcv[2])
par(mfrow=c(2, 2), mar=c(4, 5, 2, 1))
syms <- mcols(seB.filt.testing)$Symbol
myplotbcv(dgeB.filt.testing.interceptonly, "~ 1", rngbcv, syms, cutoffdisp1=10.4,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeB.filt.testing.PE, "~ Tumor", rngbcv, syms, cutoffdisp1=10.4,
          cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeB.filt.testing.PEGA, "~ Tumor + Sex", rngbcv, syms,
          cutoffdisp1=10.4, cutoffdisp2=0.8, cutoffA=9)
myplotbcv(dgeB.filt.testing.PEGASVs, "~ Tumor + Sex + SVs",
          rngbcv, syms, cutoffdisp1=10.4, cutoffdisp2=0.6, cutoffA=8)
```

```{r hypervargenesDsubset, echo=FALSE}
ord <- order(dgeB.filt.testing.PEGASVs$prior.df)
dtf <- data.frame(prior.df=dgeB.filt.testing.PEGASVs$prior.df[ord],
                  BCV=sqrt(dgeB.filt.testing.PEGASVs$tagwise.dispersion[ord]),
                  Gene=dgeB.filt.testing.PEGASVs$genes$Symbol[ord],
                  Description=dgeB.filt.testing.PEGASVs$genes$Description[ord])
ktab <- kable(dtf[1:20, ], "html", escape=FALSE, row.names=FALSE,
              caption="Top-20 hypervariable genes for the model Tumor + Sex + SVs.")
ktab <- kable_styling(ktab, bootstrap_options=c("striped", "hover", "responsive"),
                      fixed_thead=TRUE)
kable_styling(ktab, position="center")
```

# Session information

```{r, child="_session-info.Rmd"}
```

# References


