---
title: Differential expression analysis of cancer data
author:
- name: Berta Canal SimÃ³n
  affiliation:
  - &id Barcelona School of Economics
  email: berta.canal@bse.eu
- name: Adam Olivares Canal
  affiliation: *id
  email: adam.olivares@bse.eu
date: "`r format(Sys.time(), '%B %e, %Y')`"
abstract: >
  Here we perform a differential expression analysis of the filtered and normalized RNA-seq expression profiles of cancer data.
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    fig_captions: yes
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: _bibliography.bib
---

```{r setup, echo=FALSE, cache=FALSE}
library(knitr)
library(kableExtra)


knitr::opts_chunk$set(
  collapse=TRUE,
  comment="",
  fig.align="center",
  fig.wide=TRUE,
  cache=TRUE,
  cache.path="_cache/__DE_cn",
  cache.extra=R.version.string,
  autodep=TRUE
)

source(file.path("_scripts", "labelPanel.R")) ## for the labelPanel() function
source(file.path("_scripts", "myplotbcv.R")) ## for the myplotbcv() function
source(file.path("_scripts", "plotvolcano.R")) ## for the plotvolcano() function
source(file.path("_scripts", "plotma.R")) ## for the plotma() function
```

# Importing processed and filtered data

We start by importing the previously filtered and normalized RNA-seq data.

```{r, message=FALSE}
library(SummarizedExperiment)
library(edgeR)

dgeR.filt <- readRDS(file.path("_processed_data", "dgeR.filt.rds"))
seR.filt <- readRDS(file.path("_processed_data", "seR.filt.rds"))
dgeB.filt <- readRDS(file.path("_processed_data", "dgeB.filt.rds"))
seB.filt <- readRDS(file.path("_processed_data", "seB.filt.rds"))

dgeR.filt.training <- readRDS(file.path("_processed_data",
                                        "dgeR.filt.training.rds"))
seR.filt.training <- readRDS(file.path("_processed_data",
                                       "seR.filt.training.rds"))
dgeB.filt.testing <- readRDS(file.path("_processed_data",
                                       "dgeB.filt.testing.rds"))
seB.filt.testing <- readRDS(file.path("_processed_data",
                                      "seB.filt.testing.rds"))
```

# Roskams et al. (2022) data

Design matrices.

```{r, message=FALSE}
mod <- model.matrix(~ Tumor + Sex, colData(seR.filt.training))
mod0 <- model.matrix(~ Sex, colData(seR.filt.training))
```

Estimate surrogate variables (SVs).

```{r, message=FALSE}
library(sva)

IQRs <- apply(assays(seR.filt.training)$logCPM, 1, IQR)
mask <- IQRs > quantile(IQRs, prob=0.9)
sv <- sva(assays(seR.filt.training)$logCPM[mask, ],
          mod=mod, mod0=mod0)
mod <- cbind(mod, sv$sv)
```

Fit linear models using voom [@law2014voom] with sample weights. Figure
\@ref(fig:voomMtraining) shows the mean-variance trend of the data.

```{r voomMtraining, echo=FALSE, fig.height=4, fig.width=6, dpi=100, fig.cap="Mean-variance trend in the training data from Roskams et al. (2022)."}
fit.training <- voomLmFit(dgeR.filt.training, mod,
                          sample.weights=TRUE, plot=TRUE)
```

Calculate moderated _t_-statistics using an empirical Bayes procedure and
examine the extent of differential expression (DE) with an FDR < 5%.

```{r}
fit.training <- eBayes(fit.training, robust=TRUE)
res.training <- decideTests(fit.training, p.value=0.05)
summary(res.training)
```

Fetch summary DE statistics for all genes.

```{r}
tt.training <- topTable(fit.training, coef="Tumoryes", n=Inf,
                        sort.by="p")
```

Display the p-value distribution in Figure \@ref(fig:pvaldisttraining).

```{r pvaldisttraining, echo=FALSE, fig.height=5, fig.width=5, dpi=100, fig.cap="P-value distribution for the null hypothesis of no-DE."}
par(mar=c(5, 4, 1, 1))
hist(tt.training$P.Value, main="", xlab="P-value", las=1)
```

Select DE genes with FDR < 5%.

```{r}
mask <- tt.training$adj.P.Val < 0.05
DEgenes.training <- rownames(tt.training)[mask]
length(DEgenes.training)
```

Display volcano and MA plots in Figure \@ref(fig:maplottraining).

```{r maplottraining, echo=FALSE, message = FALSE, fig.height=5, fig.width=10, dpi=100, fig.cap="MA-plot training data from Roskams et al. (2022)."}
par(mfrow=c(1, 2))
plotvolcano(tt.training, highlight=DEgenes.training, FDRcutoff=0.05)
plotma(tt.training, highlight=DEgenes.training, FDRcutoff=0.05)
```

saveRDS(dgeR.filt, file=file.path("_processed_data", "dgeR.filt.rds"))
saveRDS(seR.filt, file=file.path("_processed_data", "seR.filt.rds"))

```{r}
saveRDS(DEgenes.training, file=file.path("_processed_data", "DEgenes.trainingR.rds"))
```


# Block et al. (2022) data

Design matrices.

```{r, message=FALSE}
mod <- model.matrix(~ Tumor + Sex, colData(seB.filt.testing))
mod0 <- model.matrix(~ Sex, colData(seB.filt.testing))
```

Estimate surrogate variables (SVs).

```{r, message=FALSE}
library(sva)

IQRs <- apply(assays(seB.filt.testing)$logCPM, 1, IQR)
mask <- IQRs > quantile(IQRs, prob=0.9)
sv <- sva(assays(seB.filt.testing)$logCPM[mask, ],
          mod=mod, mod0=mod0)
mod <- cbind(mod, sv$sv)
```

Fit linear models using voom [@law2014voom] with sample weights. Figure
\@ref(fig:voomMtraining) shows the mean-variance trend of the data.

```{r voomMtesting, echo=FALSE, fig.height=4, fig.width=6, dpi=100, fig.cap="Mean-variance trend in the testing data from Block et al. (2022)."}
fit.testing <- voomLmFit(dgeB.filt.testing, mod,
                         sample.weights=TRUE, plot=TRUE)
```

Calculate moderated _t_-statistics using an empirical Bayes procedure and
examine the extent of differential expression (DE) with an FDR < 5%.

```{r}
fit.testing <- eBayes(fit.testing, robust=TRUE)
res.testing <- decideTests(fit.testing, p.value=0.05)
summary(res.testing)
```

Fetch summary DE statistics for all genes.

```{r}
tt.testing <- topTable(fit.testing, coef="Tumoryes", n=Inf,
                       sort.by="p")
```

Display the p-value distribution in Figure \@ref(fig:pvaldisttesting).

```{r pvaldisttesting, echo=FALSE, fig.height=5, fig.width=5, dpi=100, fig.cap="P-value distribution for the null hypothesis of no-DE."}
par(mar=c(5, 4, 1, 1))
hist(tt.testing$P.Value, main="", xlab="P-value", las=1)
```

Select DE genes with FDR < 5%.

```{r}
mask <- tt.testing$adj.P.Val < 0.05
DEgenes.testing <- rownames(tt.testing)[mask]
length(DEgenes.testing)
```

Display MA-plot in Figure \@ref(fig:maplottesting).

```{r maplottesting, echo=FALSE, fig.height=5, fig.width=10, dpi=100, fig.cap="MA-plot testing data from Block et al. (2022)."}
par(mfrow=c(1, 2))
plotvolcano(tt.testing, highlight=DEgenes.testing, FDRcutoff=0.05)
plotma(tt.testing, highlight=DEgenes.testing, FDRcutoff=0.05)
```

# Comparison of DE genes across datasets

```{r decomparison, echo=FALSE, warning=FALSE, fig.height=6, fig.width=10, dpi=100, fig.cap="Comparison DE analyis across datasets.", eval=FALSE}
par(mfrow=c(1, 1), mar=c(5, 5, 5, 5))
traininglfc <- tt.training$logFC
names(traininglfc) <- rownames(tt.training)
testinglfc <- tt.testing$logFC
names(testinglfc) <- rownames(tt.testing)

## training vs testing
commongenes <- intersect(names(traininglfc), names(testinglfc))
plot(traininglfc[commongenes], testinglfc[commongenes], pch=19, cex=0.7,
     col="gray", xlab="", ylab="", main="", las=1, cex.lab=1.5, cex.axis=1.2)
mtext("Roskams et al. (2022) training", side=1, line=2.2, cex=1.3)
mtext(expression(paste(log[2], " fold change")), side=1, line=3.7, cex=1.3)
mtext("Block et al. (2022) testing", side=2, line=3.7, cex=1.3)
mtext(expression(paste(log[2], " fold change")), side=2, line=2.2, cex=1.3)
abline(0, 1, col="gray", lwd=2, lty=3)
abline(v=0, lwd=2, col="gray", lty=3)
abline(h=0, lwd=2, col="gray", lty=3)
rug(traininglfc, side=1, col="gray")
rug(testinglfc, side=2, col="gray")
commonDEgenes <- intersect(DEgenes.training, DEgenes.testing)
points(traininglfc[commonDEgenes], testinglfc[commonDEgenes], pch=19,
       col="darkred")
fit <- lm(testinglfc[commonDEgenes] ~ traininglfc[commonDEgenes])
abline(fit, lwd=2, col="darkred")
r2 <- sprintf("%.2f", summary(fit)$r.squared)
rhotest <- cor.test(traininglfc[commonDEgenes], testinglfc[commonDEgenes])
rho <- sprintf("%.2f", rhotest$estimate)
pv <- ifelse(summary(fit)$coefficients[2, 4] < 0.001, "0.001",
                          sprintf("%.3f", summary(fit)$coefficients[2, 4]))
rp <- vector('expression', 3)
rp[1] <- substitute(expression(italic(r) == RHO), list(RHO=rho))[2]
rp[2] <- substitute(expression(italic(R)^2 == RSQUARED), list(RSQUARED=r2))[2]
rp[3] <- substitute(expression(italic(P) < PVALUE), list(PVALUE=pv))[2]
legend("topleft", legend=rp, bty="n", text.col="darkred", cex=0.8)
```

# Session information

```{r, child="_session-info.Rmd"}
```

# References

