<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Quality assessment of preeclampsia data</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       </style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
  p.abstract{
    text-align: center;
    font-weight: bold;
  }
  div.abstract{
    margin: auto;
    width: 90%;
  }
</style>


<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="/home/adam/R/x86_64-pc-linux-gnu-library/4.4/BiocStyle/resources/html/bioconductor.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 828px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {

}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 246px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



<script>
function toggle_visibility(id1) {
  var e = document.getElementById(id1);
  e.style.display = ((e.style.display!="none") ? "none" : "block");
}
</script>

</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Biomarker Discovery in cfRNA data</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="01_qa_pe.html">QA Preeclampsia</a>
</li>
<li>
  <a href="02_de_pe.html">DE Preeclampsia</a>
</li>
<li>
  <a href="03_qa_cn.html">QA Cancer</a>
</li>
<li>
  <a href="04_de_cn.html">DE Cancer</a>
</li>
<li>
  <a href="05_pr_pe.html">Prediction preeclampsia</a>
</li>
<li>
  <a href="06_pr_cn.html">Prediction cancer</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Quality assessment of preeclampsia data</h1>
<p class="author-name">Berta Canal Simón<span class="affil-mark">1*</span> and Adam Olivares Canal<span class="affil-mark">1**</span></p>
<p class="author-affiliation"><span class="affil-mark">1</span>Barcelona School of Economics</p>
<p class="author-email"><span class="affil-mark">*</span><a href="mailto:berta.canal@bse.eu">berta.canal@bse.eu</a><br><span class="affil-mark">**</span><a href="mailto:adam.olivares@bse.eu">adam.olivares@bse.eu</a></p>
<h4 class="date">abril 29, 2024</h4>
<h4 class="abstract">Abstract</h4>
<p>Here we perform a first exploratory analysis and quality assessment of the cfRNA expression profiles of the preeclampsia data.</p>

</div>


<div id="read-molecular-and-phenotype-data" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Read molecular and phenotype data</h1>
<p>Read the molecular and phenotype data stored in serialized <code>SummarizedExperiment</code>
objects.</p>
<pre class="r"><code>library(SummarizedExperiment)

seM &lt;- readRDS(file.path(&quot;_raw_data&quot;, &quot;se.Moufarrej22.GENCODEv44.hg38.rds&quot;))
dim(seM)
[1] 37928   404
seD &lt;- readRDS(file.path(&quot;_raw_data&quot;, &quot;se.DelVecchio21.GENCODEv44.hg38.rds&quot;))
dim(seD)
[1] 37928   134</code></pre>
</div>
<div id="filtering-of-lowly-expressed-genes-and-normalization" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Filtering of lowly-expressed genes and normalization</h1>
<p>We will use the <a href="https://bioconductor.org/packages/edgeR">edgeR</a> package for
processing the raw counts and doing some exploratory analyses. We start by
building the corresponding <code>DGEList</code> objects with the corresponding row and
column data.</p>
<pre class="r"><code>library(edgeR)

dgeM &lt;- DGEList(counts=assay(seM), genes=rowData(seM), samples=colData(seM))
dgeD &lt;- DGEList(counts=assay(seD), genes=rowData(seD), samples=colData(seD))</code></pre>
<p>Filter lowly expressed genes using the function <code>filterByExpr()</code> and the
<code>Preeclampsia</code> phenotype column as grouping factor. We need to filter both,
the <code>SummarizedExperiment</code> and <code>DGEList</code> types of objects, to keep them
parallel to each other.</p>
<pre class="r"><code>mask &lt;- filterByExpr(dgeM, group=seM$Preeclampsia)
sum(mask)
[1] 5344
seM.filt &lt;- seM[mask, ]
dim(seM.filt)
[1] 5344  404
dgeM.filt &lt;- dgeM[mask, , keep.lib.sizes=FALSE]
dim(dgeM.filt)
[1] 5344  404

mask &lt;- filterByExpr(dgeD, group=seD$Preeclampsia)
sum(mask)
[1] 7313
seD.filt &lt;- seD[mask, ]
dim(seD.filt)
[1] 7313  134
dgeD.filt &lt;- dgeD[mask, , keep.lib.sizes=FALSE]
dim(dgeD.filt)
[1] 7313  134</code></pre>
<p>Calculate normalization factors on the filtered data.</p>
<pre class="r"><code>dgeM.filt &lt;- calcNormFactors(dgeM.filt)
assays(seM.filt)$logCPM &lt;- cpm(dgeM.filt, log=TRUE)

dgeD.filt &lt;- calcNormFactors(dgeD.filt)
assays(seD.filt)$logCPM &lt;- cpm(dgeD.filt, log=TRUE)</code></pre>
<p>Divide samples in the <span class="citation">(<a href="#ref-moufarrej2022early">Moufarrej et al. 2022</a>)</span> study in two subsets, one for
<em>training</em> in which we will put together the samples from the <code>Discovery</code> and
<code>Validation1</code> cohorts, and another one for <em>testing</em> formed by the samples of
the <code>Validation2</code> cohort. In both cases, we exclude samples drawn after
delivery, labeled as <code>Post-partum</code> in the <code>SamplingGAgroup</code> column.</p>
<pre class="r"><code>table(seM.filt$Cohort)

  Discovery Validation1 Validation2 
        209         106          89 
mask &lt;- seM.filt$Cohort %in% c(&quot;Discovery&quot;, &quot;Validation1&quot;) &amp;
        !seM.filt$SamplingGAgroup %in% &quot;Post-partum&quot;
seM.filt.training &lt;- seM.filt[, mask]
dim(seM.filt.training)
[1] 5344  244
dgeM.filt.training &lt;- dgeM.filt[, mask]
dim(dgeM.filt.training)
[1] 5344  244
mask &lt;- seM.filt$Cohort %in% &quot;Validation2&quot; &amp;
        !seM.filt$SamplingGAgroup %in% &quot;Post-partum&quot;
seM.filt.testing &lt;- seM.filt[, mask]
dim(seM.filt.testing)
[1] 5344   89
dgeM.filt.testing &lt;- dgeM.filt[, mask]
dim(dgeM.filt.testing)
[1] 5344   89</code></pre>
<p>From the study by <span class="citation">(<a href="#ref-delvecchio2021cell">Del Vecchio et al. 2021</a>)</span>, discard samples from non-pregnant
women, drawn after delivery, or taken from the cord blood of the neonate.</p>
<pre class="r"><code>table(seD.filt$SamplingGAgroup)

         1st Trimester          2nd Trimester          3rd Trimester 
                    25                     26                     25 
            cord blood               delivery Non pregnant Patient 1 
                    25                     26                      1 
Non pregnant Patient 2 Non pregnant Patient 3 Non pregnant Patient 4 
                     1                      1                      1 
Non pregnant Patient 5 Non pregnant Patient 6 Non pregnant Patient 7 
                     1                      1                      1 
mask &lt;- !grepl(&quot;Non pregnant&quot;, seD.filt$SamplingGAgroup) &amp;
        !seD.filt$SamplingGAgroup %in% c(&quot;delivery&quot;, &quot;cord blood&quot;)
seD.filt.subset &lt;- seD.filt[, mask]
dim(seD.filt.subset)
[1] 7313   76
dgeD.filt.subset &lt;- dgeD.filt[, mask]
dim(dgeD.filt.subset)
[1] 7313   76</code></pre>
<p>Figure <a href="#fig:mdsplot">1</a> below shows samples dissimilarity labeled by
the diagnosis of preeclampsia in each of the three datasets.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:mdsplot"></span>
<img src="01_qa_pe_files/figure-html/mdsplot-1.png" alt="Sample dissimilarity labeled by the diagnosis of preeclampsia. Multidimensional scaling (MDS) plot of sample dissimilarity for the training (a) and testing (b) data from the study by @moufarrej2022early, and for the data from the study by [@delvecchio2021cell]." width="800px"  class="widefigure" />
<p class="caption">
Figure 1: <span class="caption-title">Sample dissimilarity labeled by the diagnosis of preeclampsia</span><br>Multidimensional scaling (MDS) plot of sample dissimilarity for the training (a) and testing (b) data from the study by <span class="citation">Moufarrej et al. (<a href="#ref-moufarrej2022early">2022</a>)</span>, and for the data from the study by <span class="citation">(<a href="#ref-delvecchio2021cell">Del Vecchio et al. 2021</a>)</span>.
</p>
</div>
<p>Save <code>DGEList</code> and <code>SummarizedExperiment</code> objects including filtered data,
normalization factors and normalized log CPM units of expression.</p>
<pre class="r"><code>saveRDS(dgeM.filt, file=file.path(&quot;_processed_data&quot;, &quot;dgeM.filt.rds&quot;))
saveRDS(seM.filt, file=file.path(&quot;_processed_data&quot;, &quot;seM.filt.rds&quot;))

saveRDS(dgeD.filt, file=file.path(&quot;_processed_data&quot;, &quot;dgeD.filt.rds&quot;))
saveRDS(seD.filt, file=file.path(&quot;_processed_data&quot;, &quot;seD.filt.rds&quot;))

saveRDS(dgeM.filt.training,
        file=file.path(&quot;_processed_data&quot;, &quot;dgeM.filt.training.rds&quot;))
saveRDS(seM.filt.training,
        file=file.path(&quot;_processed_data&quot;, &quot;seM.filt.training.rds&quot;))
saveRDS(dgeM.filt.testing,
        file=file.path(&quot;_processed_data&quot;, &quot;dgeM.filt.testing.rds&quot;))
saveRDS(seM.filt.testing,
        file=file.path(&quot;_processed_data&quot;, &quot;seM.filt.testing.rds&quot;))

saveRDS(dgeD.filt.subset,
        file=file.path(&quot;_processed_data&quot;, &quot;dgeD.filt.subset.rds&quot;))
saveRDS(seD.filt.subset,
        file=file.path(&quot;_processed_data&quot;, &quot;seD.filt.subset.rds&quot;))</code></pre>
</div>
<div id="dispersion-estimates-and-hypervariable-genes" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Dispersion estimates and hypervariable genes</h1>
<p>We calculate dispersion estimates and biological coefficients of variation
<span class="citation">(<a href="#ref-mccarthy2012differential">McCarthy, Chen, and Smyth 2012</a>)</span> and use them to identify hypervariable genes for
different models of increasing complexity that will help us to select the most
sensible one for the differential expression analysis between
preeclampsia-affected and unafected mothers.</p>
<p>The four considered models are: (1) a model with the intercept term only,
(2) a model with <code>Preeclampsia</code> as main explanatory variable, (3) a model
with <code>Preeclampsia</code> as main explanatory variable and <code>SamplingGA</code> as a
covariate, and (4) a model with <code>Preeclampsia</code> as main explanatory variable,
and <code>SamplingGA</code> and surrogate variables (SVs), calculated with SVA
<span class="citation">(<a href="#ref-leek2007capturing">Leek and Storey 2007</a>)</span> as covariates. The estimation of SVs will be done using
only the expression values of the top-10% most variable genes.</p>
<div id="training-dataset-from-moufarrej-et-al.-2022" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Training dataset from Moufarrej et al. (2022)</h2>
<pre class="r"><code>library(sva)

mod &lt;- model.matrix(~ 1, colData(seM.filt.training))
dgeM.filt.training.interceptonly &lt;- estimateDisp(dgeM.filt.training, design=mod,
                                                 robust=TRUE)
summary(dgeM.filt.training.interceptonly$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  8.338   8.442   8.572   8.512   8.572   8.572 
mod &lt;- model.matrix(~ Preeclampsia, colData(seM.filt.training))
dgeM.filt.training.PE &lt;- estimateDisp(dgeM.filt.training, design=mod,
                                      robust=TRUE)
summary(dgeM.filt.training.PE$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  8.394   8.394   8.639   8.572   8.639   8.639 
mod &lt;- model.matrix(~ Preeclampsia + SamplingGA, colData(seM.filt.training))
dgeM.filt.training.PEGA &lt;- estimateDisp(dgeM.filt.training, design=mod,
                                        robust=TRUE)
summary(dgeM.filt.training.PEGA$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  8.396   8.396   8.650   8.580   8.650   8.650 
mod0 &lt;- model.matrix(~ SamplingGA, colData(seM.filt.training))
IQRs &lt;- apply(assays(seM.filt.training)$logCPM, 1, IQR)
mask &lt;- IQRs &gt; quantile(IQRs, prob=0.9)
sv &lt;- sva(assays(seM.filt.training)$logCPM[mask, ],
          mod=mod, mod0=mod0)
Number of significant surrogate variables is:  9 
Iteration (out of 5 ):1  2  3  4  5  
mod &lt;- cbind(mod, sv$sv)
dgeM.filt.training.PEGASVs &lt;- estimateDisp(dgeM.filt.training,
                                           design=mod, robust=TRUE)
summary(dgeM.filt.training.PEGASVs$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.3918 31.2056 31.2056 28.6264 31.2056 31.2056 </code></pre>
<p>Figure <a href="#fig:Mtrainingavglogcpmxvbcv">2</a> shows, for each considered model, the
biological coefficient of variation (BCV) as function of the average expression,
in log CPM units.</p>
<pre><code>Loading required package: plotrix</code></pre>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:Mtrainingavglogcpmxvbcv"></span>
<img src="01_qa_pe_files/figure-html/Mtrainingavglogcpmxvbcv-1.png" alt="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units." width="100%"  class="widefigure" />
<p class="caption">
Figure 2: <span class="caption-title">Biological coefficient of variation (BCV)</span><br>BCV values as function of the average gene expression in log-CPM units.
</p>
</div>
<p>Table <a href="#tab:hypervargenesMtraining">1</a> below shows the top-20 hypervariable genes
with lowest prior degrees of freedom for the last model.</p>
<div class='horizontal-scroll'><table class="table table-striped table-hover table-responsive table" style="color: black; margin-left: auto; margin-right: auto; color: black; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:hypervargenesMtraining">Table 1: </span>Top-20 hypervariable genes for the model Preeclampsia + SamplingGA + SVs.
</caption>
<thead>
<tr>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
prior.df
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
BCV
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Gene
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Description
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0.3917911
</td>
<td style="text-align:right;">
1.0583241
</td>
<td style="text-align:left;">
HBG1
</td>
<td style="text-align:left;">
hemoglobin subunit gamma 1
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3917932
</td>
<td style="text-align:right;">
1.1637203
</td>
<td style="text-align:left;">
HBG2
</td>
<td style="text-align:left;">
hemoglobin subunit gamma 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3917943
</td>
<td style="text-align:right;">
0.7956527
</td>
<td style="text-align:left;">
ALAS2
</td>
<td style="text-align:left;">
5’-aminolevulinate synthase 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3918679
</td>
<td style="text-align:right;">
1.5304564
</td>
<td style="text-align:left;">
C5orf64
</td>
<td style="text-align:left;">
chromosome 5 putative open reading frame 64
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3921824
</td>
<td style="text-align:right;">
0.7139674
</td>
<td style="text-align:left;">
RPS26
</td>
<td style="text-align:left;">
ribosomal protein S26
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3925444
</td>
<td style="text-align:right;">
0.6502067
</td>
<td style="text-align:left;">
IGF2
</td>
<td style="text-align:left;">
insulin like growth factor 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3926117
</td>
<td style="text-align:right;">
0.9827843
</td>
<td style="text-align:left;">
CST7
</td>
<td style="text-align:left;">
cystatin F
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3930199
</td>
<td style="text-align:right;">
0.6056745
</td>
<td style="text-align:left;">
SLC4A1
</td>
<td style="text-align:left;">
solute carrier family 4 member 1 (Diego blood group)
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3933930
</td>
<td style="text-align:right;">
0.7048308
</td>
<td style="text-align:left;">
SNORD3B-2
</td>
<td style="text-align:left;">
small nucleolar RNA, C/D box 3B-2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3984848
</td>
<td style="text-align:right;">
0.7196402
</td>
<td style="text-align:left;">
RNVU1-18
</td>
<td style="text-align:left;">
RNA, variant U1 small nuclear 18
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3984848
</td>
<td style="text-align:right;">
0.7616192
</td>
<td style="text-align:left;">
VTRNA1-2
</td>
<td style="text-align:left;">
vault RNA 1-2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3992817
</td>
<td style="text-align:right;">
0.5866055
</td>
<td style="text-align:left;">
OSBP2
</td>
<td style="text-align:left;">
oxysterol binding protein 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3992817
</td>
<td style="text-align:right;">
0.5858608
</td>
<td style="text-align:left;">
ND6
</td>
<td style="text-align:left;">
NADH dehydrogenase subunit 6
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4017543
</td>
<td style="text-align:right;">
0.6021897
</td>
<td style="text-align:left;">
H19
</td>
<td style="text-align:left;">
H19 imprinted maternally expressed transcript
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4037194
</td>
<td style="text-align:right;">
0.6662309
</td>
<td style="text-align:left;">
EGR3
</td>
<td style="text-align:left;">
early growth response 3
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4112340
</td>
<td style="text-align:right;">
0.5752273
</td>
<td style="text-align:left;">
SLC25A39
</td>
<td style="text-align:left;">
solute carrier family 25 member 39
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4132431
</td>
<td style="text-align:right;">
0.8511395
</td>
<td style="text-align:left;">
MAGI2-AS3
</td>
<td style="text-align:left;">
MAGI2 antisense RNA 3
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4141007
</td>
<td style="text-align:right;">
0.5521355
</td>
<td style="text-align:left;">
RN7SL767P
</td>
<td style="text-align:left;">
RNA, 7SL, cytoplasmic 767, pseudogene
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4161676
</td>
<td style="text-align:right;">
0.5409125
</td>
<td style="text-align:left;">
RNU1-3
</td>
<td style="text-align:left;">
RNA, U1 small nuclear 3
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4311507
</td>
<td style="text-align:right;">
0.5089252
</td>
<td style="text-align:left;">
RN7SL151P
</td>
<td style="text-align:left;">
RNA, 7SL, cytoplasmic 151, pseudogene
</td>
</tr>
</tbody>
</table></div>
</div>
<div id="testing-dataset-from-moufarrej-et-al.-2022" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Testing dataset from Moufarrej et al. (2022)</h2>
<pre class="r"><code>mod &lt;- model.matrix(~ 1, colData(seM.filt.testing))
dgeM.filt.testing.interceptonly &lt;- estimateDisp(dgeM.filt.testing, design=mod,
                                                robust=TRUE)
summary(dgeM.filt.testing.interceptonly$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  3.536   3.536   3.536   3.638   3.813   3.813 
mod &lt;- model.matrix(~ Preeclampsia, colData(seM.filt.testing))
dgeM.filt.testing.PE &lt;- estimateDisp(dgeM.filt.testing, design=mod,
                                     robust=TRUE)
summary(dgeM.filt.testing.PE$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  3.613   3.613   3.613   3.714   3.881   3.881 
mod &lt;- model.matrix(~ Preeclampsia + SamplingGA, colData(seM.filt.testing))
dgeM.filt.testing.PEGA &lt;- estimateDisp(dgeM.filt.testing, design=mod,
                                       robust=TRUE)
summary(dgeM.filt.testing.PEGA$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  3.606   3.606   3.606   3.709   3.878   3.878 
mod0 &lt;- model.matrix(~ SamplingGA, colData(seM.filt.testing))
IQRs &lt;- apply(assays(seM.filt.testing)$logCPM, 1, IQR)
mask &lt;- IQRs &gt; quantile(IQRs, prob=0.9)
sv &lt;- sva(assays(seM.filt.testing)$logCPM[mask, ],
          mod=mod, mod0=mod0)
Number of significant surrogate variables is:  4 
Iteration (out of 5 ):1  2  3  4  5  
mod &lt;- cbind(mod, sv$sv)
dgeM.filt.testing.PEGASVs &lt;- estimateDisp(dgeM.filt.testing,
                                          design=mod, robust=TRUE)
summary(dgeM.filt.testing.PEGASVs$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  5.292   7.273   7.273   7.243   7.273   7.273 </code></pre>
<p>Figure <a href="#fig:Mtestingavglogcpmxvbcv">3</a> shows, for each considered model, the
biological coefficient of variation (BCV) as function of the average expression,
n log CPM units.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:Mtestingavglogcpmxvbcv"></span>
<img src="01_qa_pe_files/figure-html/Mtestingavglogcpmxvbcv-1.png" alt="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units." width="100%"  class="widefigure" />
<p class="caption">
Figure 3: <span class="caption-title">Biological coefficient of variation (BCV)</span><br>BCV values as function of the average gene expression in log-CPM units.
</p>
</div>
</div>
<div id="subset-from-del-vecchio-et-al.-2021" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Subset from Del Vecchio et al. (2021)</h2>
<pre class="r"><code>mod &lt;- model.matrix(~ 1, colData(seD.filt.subset))
dgeD.filt.subset.interceptonly &lt;- estimateDisp(dgeD.filt.subset, design=mod,
                                               robust=TRUE)
summary(dgeD.filt.subset.interceptonly$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.4514 11.2173 11.2173 10.3761 11.2173 11.2173 
mod &lt;- model.matrix(~ Preeclampsia, colData(seD.filt.subset))
dgeD.filt.subset.PE &lt;- estimateDisp(dgeD.filt.subset, design=mod,
                                    robust=TRUE)
summary(dgeD.filt.subset.PE$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  0.413  11.531  11.531  10.688  11.531  11.531 
mod &lt;- model.matrix(~ Preeclampsia + SamplingGA, colData(seD.filt.subset))
dgeD.filt.subset.PEGA &lt;- estimateDisp(dgeD.filt.subset, design=mod,
                                      robust=TRUE)
summary(dgeD.filt.subset.PEGA$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.4149 11.7678 11.7678 10.8987 11.7678 11.7678 
mod0 &lt;- model.matrix(~ SamplingGA, colData(seD.filt.subset))
IQRs &lt;- apply(assays(seD.filt.subset)$logCPM, 1, IQR)
mask &lt;- IQRs &gt; quantile(IQRs, prob=0.9)
sv &lt;- sva(assays(seD.filt.subset)$logCPM[mask, ],
          mod=mod, mod0=mod0)
Number of significant surrogate variables is:  8 
Iteration (out of 5 ):1  2  3  4  5  
mod &lt;- cbind(mod, sv$sv)
dgeD.filt.subset.PEGASVs &lt;- estimateDisp(dgeD.filt.subset,
                                         design=mod, robust=TRUE)
summary(dgeD.filt.subset.PEGASVs$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.45   50.16   50.16   46.96   50.16   50.16 </code></pre>
<p>Figure <a href="#fig:Dsubsetavglogcpmxvbcv">4</a> shows, for each considered model, the
biological coefficient of variation (BCV) as function of the average expression,
n log CPM units.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:Dsubsetavglogcpmxvbcv"></span>
<img src="01_qa_pe_files/figure-html/Dsubsetavglogcpmxvbcv-1.png" alt="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units." width="100%"  class="widefigure" />
<p class="caption">
Figure 4: <span class="caption-title">Biological coefficient of variation (BCV)</span><br>BCV values as function of the average gene expression in log-CPM units.
</p>
</div>
<p>Table <a href="#tab:hypervargenesDsubset">2</a> below shows the top-20 hypervariable genes
with lowest prior degrees of freedom for the last model.</p>
<div class='horizontal-scroll'><table class="table table-striped table-hover table-responsive table" style="color: black; margin-left: auto; margin-right: auto; color: black; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:hypervargenesDsubset">Table 2: </span>Top-20 hypervariable genes for the model Preeclampsia + SamplingGA + SVs.
</caption>
<thead>
<tr>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
prior.df
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
BCV
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Gene
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Description
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0.4500197
</td>
<td style="text-align:right;">
0.6215532
</td>
<td style="text-align:left;">
PLAC4
</td>
<td style="text-align:left;">
placenta enriched 4
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500197
</td>
<td style="text-align:right;">
0.6716895
</td>
<td style="text-align:left;">
ALB
</td>
<td style="text-align:left;">
albumin
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500197
</td>
<td style="text-align:right;">
0.5194991
</td>
<td style="text-align:left;">
FHDC1
</td>
<td style="text-align:left;">
FH2 domain containing 1
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500197
</td>
<td style="text-align:right;">
0.4084455
</td>
<td style="text-align:left;">
TRAK2
</td>
<td style="text-align:left;">
trafficking kinesin protein 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500200
</td>
<td style="text-align:right;">
0.4612708
</td>
<td style="text-align:left;">
ALAS2
</td>
<td style="text-align:left;">
5’-aminolevulinate synthase 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500200
</td>
<td style="text-align:right;">
0.3798014
</td>
<td style="text-align:left;">
ND5
</td>
<td style="text-align:left;">
NADH dehydrogenase subunit 5
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500201
</td>
<td style="text-align:right;">
0.3725009
</td>
<td style="text-align:left;">
ND4
</td>
<td style="text-align:left;">
NADH dehydrogenase subunit 4
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500201
</td>
<td style="text-align:right;">
0.3848531
</td>
<td style="text-align:left;">
CYTB
</td>
<td style="text-align:left;">
cytochrome b
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500214
</td>
<td style="text-align:right;">
0.3783909
</td>
<td style="text-align:left;">
COX3
</td>
<td style="text-align:left;">
cytochrome c oxidase subunit III
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500219
</td>
<td style="text-align:right;">
0.3541374
</td>
<td style="text-align:left;">
COX1
</td>
<td style="text-align:left;">
cytochrome c oxidase subunit I
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500239
</td>
<td style="text-align:right;">
0.4780198
</td>
<td style="text-align:left;">
ND6
</td>
<td style="text-align:left;">
NADH dehydrogenase subunit 6
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500266
</td>
<td style="text-align:right;">
0.4469658
</td>
<td style="text-align:left;">
MTND1P23
</td>
<td style="text-align:left;">
MT-ND1 pseudogene 23
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500268
</td>
<td style="text-align:right;">
0.4087547
</td>
<td style="text-align:left;">
IGF2
</td>
<td style="text-align:left;">
insulin like growth factor 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500287
</td>
<td style="text-align:right;">
0.5637600
</td>
<td style="text-align:left;">
HBG1
</td>
<td style="text-align:left;">
hemoglobin subunit gamma 1
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500317
</td>
<td style="text-align:right;">
0.4111922
</td>
<td style="text-align:left;">
MAP3K7CL
</td>
<td style="text-align:left;">
MAP3K7 C-terminal like
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500318
</td>
<td style="text-align:right;">
0.3544274
</td>
<td style="text-align:left;">
COX2
</td>
<td style="text-align:left;">
cytochrome c oxidase subunit II
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500346
</td>
<td style="text-align:right;">
0.9197978
</td>
<td style="text-align:left;">
RPL23AP7
</td>
<td style="text-align:left;">
ribosomal protein L23a pseudogene 7
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500392
</td>
<td style="text-align:right;">
0.5096776
</td>
<td style="text-align:left;">
MTND4P12
</td>
<td style="text-align:left;">
MT-ND4 pseudogene 12
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500407
</td>
<td style="text-align:right;">
0.4149040
</td>
<td style="text-align:left;">
YOD1
</td>
<td style="text-align:left;">
YOD1 deubiquitinase
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4500407
</td>
<td style="text-align:right;">
0.3507107
</td>
<td style="text-align:left;">
BNIP3L
</td>
<td style="text-align:left;">
BCL2 interacting protein 3 like
</td>
</tr>
</tbody>
</table></div>
</div>
</div>
<div id="session-information" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Session information</h1>
<pre class="r"><code>sessionInfo()
R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=es_ES.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=es_ES.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=es_ES.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=es_ES.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Madrid
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] plotrix_3.8-2               sva_3.51.0                 
 [3] BiocParallel_1.37.1         genefilter_1.85.1          
 [5] mgcv_1.9-1                  nlme_3.1-163               
 [7] edgeR_4.1.26                limma_3.59.10              
 [9] SummarizedExperiment_1.33.3 Biobase_2.63.1             
[11] GenomicRanges_1.55.4        GenomeInfoDb_1.39.14       
[13] IRanges_2.37.1              S4Vectors_0.41.7           
[15] BiocGenerics_0.49.1         MatrixGenerics_1.15.1      
[17] matrixStats_1.3.0           kableExtra_1.4.0           
[19] knitr_1.46                  BiocStyle_2.31.0           

loaded via a namespace (and not attached):
 [1] viridisLite_0.4.0       blob_1.2.2              Biostrings_2.71.6      
 [4] fastmap_1.1.1           XML_3.99-0.9            digest_0.6.29          
 [7] lifecycle_1.0.1         survival_3.5-8          statmod_1.4.36         
[10] KEGGREST_1.43.1         RSQLite_2.2.9           magrittr_2.0.2         
[13] compiler_4.4.0          rlang_1.1.3             sass_0.4.9             
[16] tools_4.4.0             yaml_2.3.5              S4Arrays_1.3.7         
[19] bit_4.0.4               DelayedArray_0.29.9     xml2_1.3.3             
[22] abind_1.4-5             grid_4.4.0              xtable_1.8-4           
[25] colorspace_2.0-2        scales_1.1.1            tinytex_0.37           
[28] cli_3.2.0               rmarkdown_2.26          crayon_1.5.0           
[31] rstudioapi_0.13         httr_1.4.2              DBI_1.1.2              
[34] cachem_1.0.6            stringr_1.4.0           zlibbioc_1.49.3        
[37] splines_4.4.0           parallel_4.4.0          AnnotationDbi_1.65.2   
[40] BiocManager_1.30.22     XVector_0.43.1          vctrs_0.3.8            
[43] Matrix_1.6-5            jsonlite_1.7.3          bookdown_0.39          
[46] bit64_4.0.5             systemfonts_1.0.4       locfit_1.5-9.9         
[49] jquerylib_0.1.4         annotate_1.81.2         codetools_0.2-19       
[52] stringi_1.7.6           UCSC.utils_0.99.7       munsell_0.5.0          
[55] htmltools_0.5.8.1       GenomeInfoDbData_1.2.12 R6_2.5.1               
[58] evaluate_0.15           lattice_0.22-5          highr_0.9              
[61] png_0.1-7               memoise_2.0.1           bslib_0.7.0            
[64] Rcpp_1.0.12             svglite_2.1.0           SparseArray_1.3.5      
[67] xfun_0.43              </code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-delvecchio2021cell" class="csl-entry">
Del Vecchio, Giorgia, Qingjiao Li, Wenyuan Li, Shanthie Thamotharan, Anela Tosevska, Marco Morselli, Kyunghyun Sung, et al. 2021. <span>“Cell-Free DNA Methylation and Transcriptomic Signature Prediction of Pregnancies with Adverse Outcomes.”</span> <em>Epigenetics</em> 16 (6): 642–61.
</div>
<div id="ref-leek2007capturing" class="csl-entry">
Leek, Jeffrey T, and John D Storey. 2007. <span>“Capturing Heterogeneity in Gene Expression Studies by Surrogate Variable Analysis.”</span> <em>PLOS Genetics</em> 3 (9): e161.
</div>
<div id="ref-mccarthy2012differential" class="csl-entry">
McCarthy, Davis J, Yunshun Chen, and Gordon K Smyth. 2012. <span>“Differential Expression Analysis of Multifactor RNA-Seq Experiments with Respect to Biological Variation.”</span> <em>Nucleic Acids Research</em> 40 (10): 4288–97.
</div>
<div id="ref-moufarrej2022early" class="csl-entry">
Moufarrej, Mira N, Sevahn K Vorperian, Ronald J Wong, Ana A Campos, Cecele C Quaintance, Rene V Sit, Michelle Tan, et al. 2022. <span>“Early Prediction of Preeclampsia in Pregnancy with Cell-Free RNA.”</span> <em>Nature</em> 602 (7898): 689–94.
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": {
      styles: {
        ".MathJax_Display": {
           "text-align": "center",
           padding: "0px 150px 0px 65px",
           margin: "0px 0px 0.5em"
        },
        "@media screen and (max-width: 991px)": {
            ".MathJax_Display": {
               "text-align": "center",
               padding: "0 0 0 0"
            }
         }
      }
    }
  });
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<script type="text/javascript">
$(document).ready(function ()  {
  
  // Map "enter" keypress to the same action as a cursor click
  function navigateLink(e) {
    if (e.key === "Enter") {
      $(this).trigger("click");
    }
  }

  var toc_items = document.querySelectorAll(".tocify-item");
  for (var i = 0; i < toc_items.length; i++) {
    // The link role tells screen readers this is for navigation
    toc_items.item(i).setAttribute("role", "link");
    // tabindex = 0 allows selection via keyboard tab presses
    toc_items.item(i).setAttribute("tabindex", "0");
    // Listen for "Enter" keypress when item is selected
    toc_items.item(i).addEventListener("keydown", navigateLink);
  }
});
</script>

</body>
</html>
