<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Quality assessment of cancer data</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       </style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
  p.abstract{
    text-align: center;
    font-weight: bold;
  }
  div.abstract{
    margin: auto;
    width: 90%;
  }
</style>


<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="/home/adam/R/x86_64-pc-linux-gnu-library/4.4/BiocStyle/resources/html/bioconductor.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 828px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {

}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 246px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



<script>
function toggle_visibility(id1) {
  var e = document.getElementById(id1);
  e.style.display = ((e.style.display!="none") ? "none" : "block");
}
</script>

</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Biomarker Discovery in cfRNA data</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="01_qa_pe.html">QA Preeclampsia</a>
</li>
<li>
  <a href="02_de_pe.html">DE Preeclampsia</a>
</li>
<li>
  <a href="03_qa_cn.html">QA Cancer</a>
</li>
<li>
  <a href="04_de_cn.html">DE Cancer</a>
</li>
<li>
  <a href="05_pr_pe.html">Prediction preeclampsia</a>
</li>
<li>
  <a href="06_pr_cn.html">Prediction cancer</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Quality assessment of cancer data</h1>
<p class="author-name">Berta Canal Simón<span class="affil-mark">1*</span> and Adam Olivares Canal<span class="affil-mark">1**</span></p>
<p class="author-affiliation"><span class="affil-mark">1</span>Barcelona School of Economics</p>
<p class="author-email"><span class="affil-mark">*</span><a href="mailto:berta.canal@bse.eu">berta.canal@bse.eu</a><br><span class="affil-mark">**</span><a href="mailto:adam.olivares@bse.eu">adam.olivares@bse.eu</a></p>
<h4 class="date">abril 29, 2024</h4>
<h4 class="abstract">Abstract</h4>
<p>Here we perform a first exploratory analysis and quality assessment of the cfRNA expression profiles of the cancer data.</p>

</div>


<div id="read-molecular-and-phenotype-data" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Read molecular and phenotype data</h1>
<p>Read the molecular and phenotype data stored in serialized <code>SummarizedExperiment</code>
objects.</p>
<pre class="r"><code>library(SummarizedExperiment)

seR &lt;- readRDS(file.path(&quot;_raw_data&quot;, &quot;se.Roskams22.GENCODEv44.hg38.rds&quot;))
dim(seR)
[1] 37928    90
seB &lt;- readRDS(file.path(&quot;_raw_data&quot;, &quot;se.Block22.GENCODEv44.hg38.rds&quot;))
dim(seB)
[1] 37928    41</code></pre>
</div>
<div id="filtering-of-lowly-expressed-genes-and-normalization" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Filtering of lowly-expressed genes and normalization</h1>
<p>We will use the <a href="https://bioconductor.org/packages/edgeR">edgeR</a> package for
processing the raw counts and doing some exploratory analyses. We start by
building the corresponding <code>DGEList</code> objects with the corresponding row and
column data.</p>
<pre class="r"><code>library(edgeR)

dgeR &lt;- DGEList(counts=assay(seR), genes=rowData(seR), samples=colData(seR))
dgeB &lt;- DGEList(counts=assay(seB), genes=rowData(seB), samples=colData(seB))</code></pre>
<p>Filter lowly expressed genes using the function <code>filterByExpr()</code> and the
<code>Tumor</code> phenotype column as grouping factor. We need to filter both,
the <code>SummarizedExperiment</code> and <code>DGEList</code> types of objects, to keep them
parallel to each other.</p>
<p>In the article, they sequenced plasma samples of eight hepatocellular carcinoma (HCC) and ten multiple myeloma (MM) patients, 12 patients of their respective precancerous conditions, and 20 non-cancer (NC) donors. For convenience, in the project will discern between those patients with and without cancer (yes and no levels), disregarding the type of tumor.</p>
<pre class="r"><code>seR@colData@listData[[&quot;Tumor&quot;]] &lt;- factor(seR$Disease != &quot;NC&quot;, labels = c(&quot;no&quot;, &quot;yes&quot;))
dgeR$samples$Tumor &lt;- factor(seR$Disease != &quot;NC&quot;, labels = c(&quot;no&quot;, &quot;yes&quot;))

seB@colData@listData[[&quot;Tumor&quot;]] &lt;- as.factor(seB$Tumor)
dgeB$samples$Tumor &lt;- as.factor(dgeB$samples$Tumor)

mask &lt;- filterByExpr(dgeR, group=seR$Tumor)
sum(mask)
[1] 8975
seR.filt &lt;- seR[mask, ]
dim(seR.filt)
[1] 8975   90
dgeR.filt &lt;- dgeR[mask, , keep.lib.sizes=FALSE]
dim(dgeR.filt)
[1] 8975   90

mask &lt;- filterByExpr(dgeB, group=seB$Tumor)
sum(mask)
[1] 18586
seB.filt &lt;- seB[mask, ]
dim(seB.filt)
[1] 18586    41
dgeB.filt &lt;- dgeB[mask, , keep.lib.sizes=FALSE]
dim(dgeB.filt)
[1] 18586    41</code></pre>
<p>Calculate normalization factors on the filtered data.</p>
<pre class="r"><code>dgeR.filt &lt;- calcNormFactors(dgeR.filt)
assays(seR.filt, withDimnames=FALSE)$logCPM &lt;- cpm(dgeR.filt, log=TRUE)

dgeB.filt &lt;- calcNormFactors(dgeB.filt)
assays(seB.filt)$logCPM &lt;- cpm(dgeB.filt, log=TRUE)</code></pre>
<p>Fo simplicity, change the name of the data from articles <span class="citation">(<a href="#ref-Roskams2022">Roskams-Hieter et al. 2022</a>)</span> and [PLACEHOLDER] to <em>training</em> and <em>testing</em>, respectively.</p>
<pre class="r"><code>seR.filt.training &lt;- seR.filt
dgeR.filt.training &lt;- dgeR.filt

seB.filt.testing &lt;- seB.filt
dgeB.filt.testing &lt;- dgeB.filt</code></pre>
<p>Figure <a href="#fig:mdsplot">1</a> below shows samples dissimilarity labeled by
the diagnosis of cancer in each of the two datasets.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:mdsplot"></span>
<img src="03_qa_cn_files/figure-html/mdsplot-1.png" alt="Sample dissimilarity labeled by the diagnosis of cancer. Multidimensional scaling (MDS) plot of sample dissimilarity for the training (a) and testing (b) data from the study by @Roskams2022, and for the data from the study by [PLACEHOLDER]." width="800px"  class="widefigure" />
<p class="caption">
Figure 1: <span class="caption-title">Sample dissimilarity labeled by the diagnosis of cancer</span><br>Multidimensional scaling (MDS) plot of sample dissimilarity for the training (a) and testing (b) data from the study by <span class="citation">Roskams-Hieter et al. (<a href="#ref-Roskams2022">2022</a>)</span>, and for the data from the study by [PLACEHOLDER].
</p>
</div>
<p>Save <code>DGEList</code> and <code>SummarizedExperiment</code> objects including filtered data,
normalization factors and normalized log CPM units of expression.</p>
</div>
<div id="data-imputation-provisional" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Data imputation (PROVISIONAL)</h1>
<div id="testing-dataset" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Testing dataset</h2>
<p>Impute missing Age values by mean.</p>
<pre class="r"><code>mean.ageB &lt;- mean(dgeB.filt.testing$samples$Age, na.rm = TRUE)
is.na.ageB &lt;- which(is.na(dgeB.filt.testing$samples$Age) == TRUE)
dgeB.filt.testing$samples$Age[is.na.ageB] &lt;- mean.ageB
seB.filt.testing$Age[is.na.ageB] &lt;- mean.ageB</code></pre>
<p>The estimation of missing data for <code>Sex</code> covariate uses a KNN imputer based on the counts for Y-linked genes (genes located on the Y chromosome). Although no Y-linked genes counts should be present in female samples, it does not hold for all samples in our dataset (maybe due to technical reasons). This led to disregarding the criteria of imputing based on the assignment of male gender to those individuals with count different than 0 for the Y-linked genes.</p>
<p>Impute missing values of Sex variable with a KNN imputer (k = 5).</p>
<pre class="r"><code>mask &lt;- grepl(&quot;Y-linked&quot;, dgeB.filt.testing$genes$Description)
counts_Y_linked &lt;- dgeB.filt.testing$counts[mask,]
counts_Y_linked &lt;- as.data.frame(t(counts_Y_linked))
counts_Y_linked$Sex &lt;- as.factor(dgeB.filt.testing$samples$Sex)
data_imputed &lt;- kNN(counts_Y_linked, k=5, variable=&quot;Sex&quot;)

dgeB.filt.testing$samples$Sex &lt;- data_imputed$Sex
seB.filt.testing$Sex &lt;- data_imputed$Sex</code></pre>
<pre class="r"><code>saveRDS(dgeR.filt, file=file.path(&quot;_processed_data&quot;, &quot;dgeR.filt.rds&quot;))
saveRDS(seR.filt, file=file.path(&quot;_processed_data&quot;, &quot;seR.filt.rds&quot;))

saveRDS(dgeB.filt, file=file.path(&quot;_processed_data&quot;, &quot;dgeB.filt.rds&quot;))
saveRDS(seB.filt, file=file.path(&quot;_processed_data&quot;, &quot;seB.filt.rds&quot;))

saveRDS(dgeR.filt.training,
        file=file.path(&quot;_processed_data&quot;, &quot;dgeR.filt.training.rds&quot;))
saveRDS(seR.filt.training,
        file=file.path(&quot;_processed_data&quot;, &quot;seR.filt.training.rds&quot;))
saveRDS(dgeB.filt.testing,
        file=file.path(&quot;_processed_data&quot;, &quot;dgeB.filt.testing.rds&quot;))
saveRDS(seB.filt.testing,
        file=file.path(&quot;_processed_data&quot;, &quot;seB.filt.testing.rds&quot;))</code></pre>
</div>
</div>
<div id="dispersion-estimates-and-hypervariable-genes" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Dispersion estimates and hypervariable genes</h1>
<p>We calculate dispersion estimates and biological coefficients of variation
<span class="citation">(<a href="#ref-mccarthy2012differential">McCarthy, Chen, and Smyth 2012</a>)</span> and use them to identify hypervariable genes for
different models of increasing complexity that will help us to select the most
sensible one for the differential expression analysis between
Cancer-affected and unafected individuals.</p>
<p>The four considered models are: (1) a model with the intercept term only,
(2) a model with <code>Cancer</code> as main explanatory variable, (3) a model
with <code>Cancer</code> as main explanatory variable and <code>Age</code> and <code>Sex</code> as
covariates, and (4) a model with <code>Cancer</code> as main explanatory variable,
and <code>Age</code>, <code>Sex</code> and surrogate variables (SVs), calculated with SVA
<span class="citation">(<a href="#ref-leek2007capturing">Leek and Storey 2007</a>)</span> as covariates. The estimation of SVs will be done using
only the expression values of the top-10% most variable genes.</p>
<div id="training-dataset" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Training dataset</h2>
<pre class="r"><code>library(sva)

mod &lt;- model.matrix(~ 1, colData(seR.filt.training))
dgeR.filt.training.interceptonly &lt;- estimateDisp(dgeR.filt.training, design=mod,
                                                 robust=TRUE)
summary(dgeR.filt.training.interceptonly$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  4.588   5.082   5.082   4.999   5.082   5.082 

mod &lt;- model.matrix(~ Tumor, colData(seR.filt.training))
dgeR.filt.training.PE &lt;- estimateDisp(dgeR.filt.training, design=mod,
                                      robust=TRUE)
summary(dgeR.filt.training.PE$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  4.630   5.125   5.125   5.042   5.125   5.125 

mod &lt;- model.matrix(~ Tumor + Age + Sex, colData(seR.filt.training))
dgeR.filt.training.PEGA &lt;- estimateDisp(dgeR.filt.training, design=mod,
                                        robust=TRUE)
summary(dgeR.filt.training.PEGA$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  4.680   5.181   5.181   5.099   5.181   5.181 

mod0 &lt;- model.matrix(~ Age + Sex, colData(seR.filt.training))
IQRs &lt;- apply(assays(seR.filt.training)$logCPM, 1, IQR)
mask &lt;- IQRs &gt; quantile(IQRs, prob=0.9)
sv &lt;- sva(assays(seR.filt.training)$logCPM[mask, ],
          mod=mod, mod0=mod0)
Number of significant surrogate variables is:  4 
Iteration (out of 5 ):1  2  3  4  5  
mod &lt;- cbind(mod, sv$sv)
dgeR.filt.training.PEGASVs &lt;- estimateDisp(dgeR.filt.training,
                                           design=mod, robust=TRUE)
summary(dgeR.filt.training.PEGASVs$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.3216 11.4218 11.4218 10.7224 11.4218 11.4218 </code></pre>
<p>Figure <a href="#fig:Mtrainingavglogcpmxvbcv">2</a> shows, for each considered model, the
biological coefficient of variation (BCV) as function of the average expression,
in log CPM units.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:Mtrainingavglogcpmxvbcv"></span>
<img src="03_qa_cn_files/figure-html/Mtrainingavglogcpmxvbcv-1.png" alt="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units." width="100%"  class="widefigure" />
<p class="caption">
Figure 2: <span class="caption-title">Biological coefficient of variation (BCV)</span><br>BCV values as function of the average gene expression in log-CPM units.
</p>
</div>
<p>Table <a href="#tab:hypervargenesMtrainingcn">1</a> below shows the top-20 hypervariable genes
with lowest prior degrees of freedom for the last model.</p>
<div class='horizontal-scroll'><table class="table table-striped table-hover table-responsive table" style="color: black; margin-left: auto; margin-right: auto; color: black; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:hypervargenesMtrainingcn">Table 1: </span>Top-20 hypervariable genes for the model Tumor + Sex + Age + SVs.
</caption>
<thead>
<tr>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
prior.df
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
BCV
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Gene
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Description
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0.3215709
</td>
<td style="text-align:right;">
1.2545774
</td>
<td style="text-align:left;">
RPL23AP7
</td>
<td style="text-align:left;">
ribosomal protein L23a pseudogene 7
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3215709
</td>
<td style="text-align:right;">
1.2071335
</td>
<td style="text-align:left;">
HBG1
</td>
<td style="text-align:left;">
hemoglobin subunit gamma 1
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3215709
</td>
<td style="text-align:right;">
1.2908075
</td>
<td style="text-align:left;">
HBG2
</td>
<td style="text-align:left;">
hemoglobin subunit gamma 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3282184
</td>
<td style="text-align:right;">
1.3357218
</td>
<td style="text-align:left;">
RPL23AP25
</td>
<td style="text-align:left;">
ribosomal protein L23a pseudogene 25
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4036953
</td>
<td style="text-align:right;">
0.9633883
</td>
<td style="text-align:left;">
CENPBD2P
</td>
<td style="text-align:left;">
CENPB DNA-binding domains containing 2, pseudogene
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4036953
</td>
<td style="text-align:right;">
0.8382717
</td>
<td style="text-align:left;">
ALAS2
</td>
<td style="text-align:left;">
5’-aminolevulinate synthase 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4187385
</td>
<td style="text-align:right;">
1.0862159
</td>
<td style="text-align:left;">
IGFBP2
</td>
<td style="text-align:left;">
insulin like growth factor binding protein 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.6606593
</td>
<td style="text-align:right;">
1.7116389
</td>
<td style="text-align:left;">
FLG
</td>
<td style="text-align:left;">
filaggrin
</td>
</tr>
<tr>
<td style="text-align:right;">
0.6606593
</td>
<td style="text-align:right;">
1.3351493
</td>
<td style="text-align:left;">
RNF182
</td>
<td style="text-align:left;">
ring finger protein 182
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7146786
</td>
<td style="text-align:right;">
0.7104536
</td>
<td style="text-align:left;">
CENPF
</td>
<td style="text-align:left;">
centromere protein F
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7146786
</td>
<td style="text-align:right;">
0.9016610
</td>
<td style="text-align:left;">
PIGC
</td>
<td style="text-align:left;">
phosphatidylinositol glycan anchor biosynthesis class C
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7146786
</td>
<td style="text-align:right;">
1.6381415
</td>
<td style="text-align:left;">
PPP1R3B-DT
</td>
<td style="text-align:left;">
PPP1R3B divergent transcript
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7146786
</td>
<td style="text-align:right;">
0.9793806
</td>
<td style="text-align:left;">
IGHG1
</td>
<td style="text-align:left;">
immunoglobulin heavy constant gamma 1 (G1m marker)
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7311274
</td>
<td style="text-align:right;">
0.7691979
</td>
<td style="text-align:left;">
MTND4P12
</td>
<td style="text-align:left;">
MT-ND4 pseudogene 12
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7311274
</td>
<td style="text-align:right;">
0.7727424
</td>
<td style="text-align:left;">
MAGI2-AS3
</td>
<td style="text-align:left;">
MAGI2 antisense RNA 3
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7311274
</td>
<td style="text-align:right;">
0.8118646
</td>
<td style="text-align:left;">
TMEM70
</td>
<td style="text-align:left;">
transmembrane protein 70
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7341925
</td>
<td style="text-align:right;">
0.7093576
</td>
<td style="text-align:left;">
ALB
</td>
<td style="text-align:left;">
albumin
</td>
</tr>
<tr>
<td style="text-align:right;">
0.7341925
</td>
<td style="text-align:right;">
0.6790055
</td>
<td style="text-align:left;">
FHDC1
</td>
<td style="text-align:left;">
FH2 domain containing 1
</td>
</tr>
<tr>
<td style="text-align:right;">
0.8141450
</td>
<td style="text-align:right;">
0.7674031
</td>
<td style="text-align:left;">
GCOM1
</td>
<td style="text-align:left;">
GCOM1, MYZAP-POLR2M combined locus
</td>
</tr>
<tr>
<td style="text-align:right;">
0.8141450
</td>
<td style="text-align:right;">
1.0816244
</td>
<td style="text-align:left;">
XACT
</td>
<td style="text-align:left;">
X active specific transcript
</td>
</tr>
</tbody>
</table></div>
</div>
<div id="testing-data" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Testing data</h2>
<pre class="r"><code>mod &lt;- model.matrix(~ 1, colData(seB.filt.testing))
dgeB.filt.testing.interceptonly &lt;- estimateDisp(dgeB.filt.testing, design=mod,
                                                robust=TRUE)
summary(dgeB.filt.testing.interceptonly$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  9.006   9.776   9.776   9.759   9.776   9.776 
mod &lt;- model.matrix(~ Tumor, colData(seB.filt.testing))
dgeB.filt.testing.PE &lt;- estimateDisp(dgeB.filt.testing, design=mod,
                                     robust=TRUE)
summary(dgeB.filt.testing.PE$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  9.956   9.956   9.956  10.090  10.181  10.517 
mod &lt;- model.matrix(~ Tumor + Age + Sex, colData(seB.filt.testing))
dgeB.filt.testing.PEGA &lt;- estimateDisp(dgeB.filt.testing, design=mod,
                                       robust=TRUE)
summary(dgeB.filt.testing.PEGA$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  10.50   10.50   10.50   10.65   10.76   11.11 
mod0 &lt;- model.matrix(~ Age + Sex, colData(seB.filt.testing))
IQRs &lt;- apply(assays(seB.filt.testing)$logCPM, 1, IQR)
mask &lt;- IQRs &gt; quantile(IQRs, prob=0.9)
sv &lt;- sva(assays(seB.filt.testing)$logCPM[mask, ],
          mod=mod, mod0=mod0)
Number of significant surrogate variables is:  3 
Iteration (out of 5 ):1  2  3  4  5  
mod &lt;- cbind(mod, sv$sv)
dgeB.filt.testing.PEGASVs &lt;- estimateDisp(dgeB.filt.testing,
                                          design=mod, robust=TRUE)
summary(dgeB.filt.testing.PEGASVs$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  4.344  11.407  11.407  11.064  11.407  11.407 </code></pre>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:Mtestingavglogcpmxvbcv"></span>
<img src="03_qa_cn_files/figure-html/Mtestingavglogcpmxvbcv-1.png" alt="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units." width="100%"  class="widefigure" />
<p class="caption">
Figure 3: <span class="caption-title">Biological coefficient of variation (BCV)</span><br>BCV values as function of the average gene expression in log-CPM units.
</p>
</div>
<div class='horizontal-scroll'><table class="table table-striped table-hover table-responsive table" style="color: black; margin-left: auto; margin-right: auto; color: black; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:hypervargenesDsubset">Table 2: </span>Top-20 hypervariable genes for the model Tumor + Age + Sex + SVs.
</caption>
<thead>
<tr>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
prior.df
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
BCV
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Gene
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Description
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.371179
</td>
<td style="text-align:left;">
RNU1-4
</td>
<td style="text-align:left;">
RNA, U1 small nuclear 4
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.271210
</td>
<td style="text-align:left;">
ANGPTL3
</td>
<td style="text-align:left;">
angiopoietin like 3
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.618190
</td>
<td style="text-align:left;">
APCS
</td>
<td style="text-align:left;">
amyloid P component, serum
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.244504
</td>
<td style="text-align:left;">
FMO3
</td>
<td style="text-align:left;">
flavin containing dimethylaniline monoxygenase 3
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.588583
</td>
<td style="text-align:left;">
PRG4
</td>
<td style="text-align:left;">
proteoglycan 4
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.274640
</td>
<td style="text-align:left;">
CFH
</td>
<td style="text-align:left;">
complement factor H
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.427355
</td>
<td style="text-align:left;">
CFHR4
</td>
<td style="text-align:left;">
complement factor H related 4
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.300903
</td>
<td style="text-align:left;">
CFHR2
</td>
<td style="text-align:left;">
complement factor H related 2
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.614350
</td>
<td style="text-align:left;">
C4BPA
</td>
<td style="text-align:left;">
complement component 4 binding protein alpha
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.154461
</td>
<td style="text-align:left;">
EPHX1
</td>
<td style="text-align:left;">
epoxide hydrolase 1
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.379213
</td>
<td style="text-align:left;">
C8B
</td>
<td style="text-align:left;">
complement C8 beta chain
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.548003
</td>
<td style="text-align:left;">
HMGCS2
</td>
<td style="text-align:left;">
3-hydroxy-3-methylglutaryl-CoA synthase 2
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.225921
</td>
<td style="text-align:left;">
PDZK1
</td>
<td style="text-align:left;">
PDZ domain containing 1
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.314657
</td>
<td style="text-align:left;">
HJV
</td>
<td style="text-align:left;">
hemojuvelin BMP co-receptor
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.242095
</td>
<td style="text-align:left;">
PKLR
</td>
<td style="text-align:left;">
pyruvate kinase L/R
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.547399
</td>
<td style="text-align:left;">
CRP
</td>
<td style="text-align:left;">
C-reactive protein
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.322220
</td>
<td style="text-align:left;">
APOA2
</td>
<td style="text-align:left;">
apolipoprotein A2
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.251799
</td>
<td style="text-align:left;">
F5
</td>
<td style="text-align:left;">
coagulation factor V
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.318453
</td>
<td style="text-align:left;">
SERPINC1
</td>
<td style="text-align:left;">
serpin family C member 1
</td>
</tr>
<tr>
<td style="text-align:right;">
4.34368
</td>
<td style="text-align:right;">
1.436629
</td>
<td style="text-align:left;">
CHI3L1
</td>
<td style="text-align:left;">
chitinase 3 like 1
</td>
</tr>
</tbody>
</table></div>
</div>
</div>
<div id="session-information" class="section level1" number="5">
<h1><span class="header-section-number">5</span> Session information</h1>
<pre class="r"><code>sessionInfo()
R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=es_ES.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=es_ES.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=es_ES.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=es_ES.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Madrid
tzcode source: system (glibc)

attached base packages:
[1] stats4    grid      stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] plotrix_3.8-2               sva_3.51.0                 
 [3] BiocParallel_1.37.1         genefilter_1.85.1          
 [5] mgcv_1.9-1                  nlme_3.1-163               
 [7] edgeR_4.1.26                limma_3.59.10              
 [9] SummarizedExperiment_1.33.3 Biobase_2.63.1             
[11] GenomicRanges_1.55.4        GenomeInfoDb_1.39.14       
[13] IRanges_2.37.1              S4Vectors_0.41.7           
[15] BiocGenerics_0.49.1         MatrixGenerics_1.15.1      
[17] matrixStats_1.3.0           VIM_6.2.2                  
[19] colorspace_2.0-2            kableExtra_1.4.0           
[21] knitr_1.46                  BiocStyle_2.31.0           

loaded via a namespace (and not attached):
 [1] DBI_1.1.2               rlang_1.1.3             magrittr_2.0.2         
 [4] e1071_1.7-9             compiler_4.4.0          RSQLite_2.2.9          
 [7] png_0.1-7               systemfonts_1.0.4       vctrs_0.3.8            
[10] stringr_1.4.0           crayon_1.5.0            fastmap_1.1.1          
[13] XVector_0.43.1          rmarkdown_2.26          UCSC.utils_0.99.7      
[16] tinytex_0.37            bit_4.0.4               xfun_0.43              
[19] zlibbioc_1.49.3         cachem_1.0.6            jsonlite_1.7.3         
[22] blob_1.2.2              highr_0.9               DelayedArray_0.29.9    
[25] parallel_4.4.0          R6_2.5.1                bslib_0.7.0            
[28] stringi_1.7.6           vcd_1.4-12              ranger_0.16.0          
[31] car_3.1-2               boot_1.3-30             lmtest_0.9-40          
[34] jquerylib_0.1.4         Rcpp_1.0.12             bookdown_0.39          
[37] zoo_1.8-9               Matrix_1.6-5            splines_4.4.0          
[40] nnet_7.3-19             rstudioapi_0.13         abind_1.4-5            
[43] yaml_2.3.5              codetools_0.2-19        lattice_0.22-5         
[46] KEGGREST_1.43.1         evaluate_0.15           survival_3.5-8         
[49] proxy_0.4-26            xml2_1.3.3              Biostrings_2.71.6      
[52] BiocManager_1.30.22     carData_3.0-5           sp_1.4-6               
[55] munsell_0.5.0           scales_1.1.1            laeken_0.5.3           
[58] xtable_1.8-4            class_7.3-22            tools_4.4.0            
[61] robustbase_0.93-9       data.table_1.14.2       annotate_1.81.2        
[64] locfit_1.5-9.9          XML_3.99-0.9            AnnotationDbi_1.65.2   
[67] GenomeInfoDbData_1.2.12 cli_3.2.0               S4Arrays_1.3.7         
[70] viridisLite_0.4.0       svglite_2.1.0           DEoptimR_1.0-10        
[73] sass_0.4.9              digest_0.6.29           SparseArray_1.3.5      
[76] memoise_2.0.1           htmltools_0.5.8.1       lifecycle_1.0.1        
[79] httr_1.4.2              statmod_1.4.36          bit64_4.0.5            
[82] MASS_7.3-60            </code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-leek2007capturing" class="csl-entry">
Leek, Jeffrey T, and John D Storey. 2007. <span>“Capturing Heterogeneity in Gene Expression Studies by Surrogate Variable Analysis.”</span> <em>PLOS Genetics</em> 3 (9): e161.
</div>
<div id="ref-mccarthy2012differential" class="csl-entry">
McCarthy, Davis J, Yunshun Chen, and Gordon K Smyth. 2012. <span>“Differential Expression Analysis of Multifactor RNA-Seq Experiments with Respect to Biological Variation.”</span> <em>Nucleic Acids Research</em> 40 (10): 4288–97.
</div>
<div id="ref-Roskams2022" class="csl-entry">
Roskams-Hieter, Breeshey, Hyun Ji Kim, Pavana Anur, Josiah T. Wagner, Rowan Callahan, Elias Spiliotopoulos, Charles Ward Kirschbaum, et al. 2022. <span>“Plasma Cell-Free RNA Profiling Distinguishes Cancers from Pre-Malignant Conditions in Solid and Hematologic Malignancies.”</span> <em>Npj Precision Oncology</em> 6. <a href="https://doi.org/10.1038/s41698-022-00270-y">https://doi.org/10.1038/s41698-022-00270-y</a>.
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": {
      styles: {
        ".MathJax_Display": {
           "text-align": "center",
           padding: "0px 150px 0px 65px",
           margin: "0px 0px 0.5em"
        },
        "@media screen and (max-width: 991px)": {
            ".MathJax_Display": {
               "text-align": "center",
               padding: "0 0 0 0"
            }
         }
      }
    }
  });
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<script type="text/javascript">
$(document).ready(function ()  {
  
  // Map "enter" keypress to the same action as a cursor click
  function navigateLink(e) {
    if (e.key === "Enter") {
      $(this).trigger("click");
    }
  }

  var toc_items = document.querySelectorAll(".tocify-item");
  for (var i = 0; i < toc_items.length; i++) {
    // The link role tells screen readers this is for navigation
    toc_items.item(i).setAttribute("role", "link");
    // tabindex = 0 allows selection via keyboard tab presses
    toc_items.item(i).setAttribute("tabindex", "0");
    // Listen for "Enter" keypress when item is selected
    toc_items.item(i).addEventListener("keydown", navigateLink);
  }
});
</script>

</body>
</html>
