<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-05-08" />

<title>Quality assessment of cancer data</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       </style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
  p.abstract{
    text-align: center;
    font-weight: bold;
  }
  div.abstract{
    margin: auto;
    width: 90%;
  }
</style>


<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/BiocStyle/resources/html/bioconductor.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 828px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {

}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 246px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



<script>
function toggle_visibility(id1) {
  var e = document.getElementById(id1);
  e.style.display = ((e.style.display!="none") ? "none" : "block");
}
</script>

</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Biomarker Discovery in cfRNA data</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="01_qa_pe.html">QA Preeclampsia</a>
</li>
<li>
  <a href="02_de_pe.html">DE Preeclampsia</a>
</li>
<li>
  <a href="03_qa_cn.html">QA Cancer</a>
</li>
<li>
  <a href="04_de_cn.html">DE Cancer</a>
</li>
<li>
  <a href="05_pr_pe.html">Prediction preeclampsia</a>
</li>
<li>
  <a href="06_pr_cn.html">Prediction cancer</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Quality assessment of cancer data</h1>
<p class="author-name">Berta Canal Simón<span class="affil-mark">1*</span> and Adam Olivares Canal<span class="affil-mark">1**</span></p>
<p class="author-affiliation"><span class="affil-mark">1</span>Barcelona School of Economics</p>
<p class="author-email"><span class="affil-mark">*</span><a href="mailto:berta.canal@bse.eu">berta.canal@bse.eu</a><br><span class="affil-mark">**</span><a href="mailto:adam.olivares@bse.eu">adam.olivares@bse.eu</a></p>
<h4 class="date">May 8, 2024</h4>
<h4 class="abstract">Abstract</h4>
<p>Here we perform a first exploratory analysis and quality assessment of the cfRNA expression profiles of the cancer data.</p>

</div>


<div id="read-molecular-and-phenotype-data" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Read molecular and phenotype data</h1>
<p>Read the molecular and phenotype data stored in serialized <code>SummarizedExperiment</code>
objects.</p>
<pre class="r"><code>library(SummarizedExperiment)

seR &lt;- readRDS(file.path(&quot;_raw_data&quot;, &quot;se.Roskams22.GENCODEv44.hg38.rds&quot;))
dim(seR)
[1] 37928    90
seB &lt;- readRDS(file.path(&quot;_raw_data&quot;, &quot;se.Block22.GENCODEv44.hg38.rds&quot;))
dim(seB)
[1] 37928    41</code></pre>
</div>
<div id="filtering-of-lowly-expressed-genes-and-normalization" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Filtering of lowly-expressed genes and normalization</h1>
<p>In <span class="citation">(<a href="#ref-Roskams2022">Roskams-Hieter et al. 2022</a>)</span>, they sequenced plasma samples of eight hepatocellular carcinoma (HCC) and ten multiple myeloma (MM) patients, 12 patients of their respective precancerous conditions, and 20 non-cancer (NC) donors. To avoid within group heterogeneity, our study focuses on individuals with hepatocellular carcinoma (HCC) and non-cancer (NC) for the control. Otherwise, our analysis could increase between group variance and thus lead to spurious results.</p>
<p>NOTA: See abstract from the paper and suplementary tables to modify this paragraph.</p>
<p>For <span class="citation">(<a href="#ref-Block2022">Block et al. 2022</a>)</span>, the project also focus on individuals with hepatocellular carcinoma (HCC) in plasma (i.e. with Microvesicle == ‘no’) and non-cancer (NC) for the control.</p>
<pre class="r"><code>table(colData(seR)$Disease)

Cirr  HCC MGUS   MM   NC 
   4   28    9   19   30 
mask &lt;- colData(seR)$Disease %in% c(&#39;NC&#39;, &#39;HCC&#39;)
seR.subset &lt;- seR[,mask]
dim(seR.subset)
[1] 37928    58
colData(seR.subset)$Tumor &lt;- factor(seR.subset$Disease == &#39;HCC&#39;, labels = c(&#39;no&#39;, &#39;yes&#39;))</code></pre>
<pre class="r"><code>table(seB$SampleType)

HCC LCS NHC NLS 
 25   9   6   1 
table(seB$SampleType, seB$Tissue)
     
      normal_liver plasma plasma_vesicles tumor_liver
  HCC            0     14               5           6
  LCS            0      9               0           0
  NHC            0      6               0           0
  NLS            1      0               0           0
mask &lt;- (seB$SampleType == &quot;HCC&quot; &amp;
         (seB$Tissue %in% c(&quot;plasma&quot;, &quot;plasma_vesicles&quot;))) |
        seB$SampleType == &quot;NHC&quot;
seB.subset &lt;- seB[,mask]
dim(seB.subset)
[1] 37928    25
colData(seB.subset)$Tumor &lt;- factor(seB.subset$SampleType == &#39;HCC&#39;,
                                    labels = c(&#39;no&#39;, &#39;yes&#39;))
table(seB.subset$SampleType)

HCC NHC 
 19   6 </code></pre>
<p>We will use the <a href="https://bioconductor.org/packages/edgeR">edgeR</a> package for
processing the raw counts and doing some exploratory analyses. We start by
building the corresponding <code>DGEList</code> objects with the corresponding row and
column data.</p>
<pre class="r"><code>library(edgeR)

dgeR &lt;- DGEList(counts=assay(seR.subset), genes=rowData(seR.subset),
                samples=colData(seR.subset))
dgeB &lt;- DGEList(counts=assay(seB.subset), genes=rowData(seB.subset),
                samples=colData(seB.subset))</code></pre>
<p>Filter lowly expressed genes using the function <code>filterByExpr()</code> and the
<code>Tumor</code> phenotype column as grouping factor. We need to filter both,
the <code>SummarizedExperiment</code> and <code>DGEList</code> types of objects, to keep them
parallel to each other.</p>
<pre class="r"><code>mask &lt;- filterByExpr(dgeR, group=seR.subset$Tumor)
sum(mask)
[1] 8387
seR.filt &lt;- seR.subset[mask, ]
dim(seR.filt)
[1] 8387   58
dgeR.filt &lt;- dgeR[mask, , keep.lib.sizes=FALSE]
dim(dgeR.filt)
[1] 8387   58

mask &lt;- filterByExpr(dgeB, group=seB.subset$Tumor)
sum(mask)
[1] 15613
seB.filt &lt;- seB.subset[mask, ]
dim(seB.filt)
[1] 15613    25
dgeB.filt &lt;- dgeB[mask, , keep.lib.sizes=FALSE]
dim(dgeB.filt)
[1] 15613    25</code></pre>
<p>Calculate normalization factors on the filtered data.</p>
<pre class="r"><code>dgeR.filt &lt;- calcNormFactors(dgeR.filt)
assays(seR.filt, withDimnames=FALSE)$logCPM &lt;- cpm(dgeR.filt, log=TRUE)

dgeB.filt &lt;- calcNormFactors(dgeB.filt)
assays(seB.filt)$logCPM &lt;- cpm(dgeB.filt, log=TRUE)</code></pre>
<p>Fo simplicity, change the name of the data from articles <span class="citation">(<a href="#ref-Roskams2022">Roskams-Hieter et al. 2022</a>)</span> and <span class="citation">(<a href="#ref-Block2022">Block et al. 2022</a>)</span> to <em>training</em> and <em>testing</em>, respectively.</p>
<pre class="r"><code>seR.filt.training &lt;- seR.filt
dgeR.filt.training &lt;- dgeR.filt

seB.filt.testing &lt;- seB.filt
dgeB.filt.testing &lt;- dgeB.filt</code></pre>
<p>Figure <a href="#fig:mdsplot">1</a> below shows samples dissimilarity labeled by
the diagnosis of cancer in each of the two datasets.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:mdsplot"></span>
<img src="03_qa_cn_files/figure-html/mdsplot-1.png" alt="Sample dissimilarity labeled by the diagnosis of cancer. Multidimensional scaling (MDS) plot of sample dissimilarity for the training (a) and testing (b) data from the study by @Roskams2022, and for the data from the study by @Block2022." width="800px"  class="widefigure" />
<p class="caption">
Figure 1: <span class="caption-title">Sample dissimilarity labeled by the diagnosis of cancer</span><br>Multidimensional scaling (MDS) plot of sample dissimilarity for the training (a) and testing (b) data from the study by <span class="citation">Roskams-Hieter et al. (<a href="#ref-Roskams2022">2022</a>)</span>, and for the data from the study by <span class="citation">Block et al. (<a href="#ref-Block2022">2022</a>)</span>.
</p>
</div>
</div>
<div id="data-imputation" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Data imputation</h1>
<div id="testing-dataset" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Testing dataset</h2>
<div id="variable-age" class="section level3" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Variable ‘Age’</h3>
<p>Impute missing age values by mean.</p>
<pre class="r"><code>mean.ageB &lt;- mean(seB.filt.testing$Age, na.rm = TRUE)
mask &lt;- is.na(seB.filt.testing$Age)
seB.filt.testing$Age[mask] &lt;- round(mean.ageB,3)</code></pre>
</div>
<div id="variable-sex" class="section level3" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> Variable ‘Sex’</h3>
<p>The estimation of missing data for <code>Sex</code> covariate uses a KNN imputer (k = 5) based on the counts for MSY genes (genes from the male-specific region of the Y chromosome). Although no MSY genes counts should be present in female samples, it does not hold for all samples in our dataset (maybe due to technical reasons). This led to disregarding the criteria of imputing based on the assignment of male gender to those individuals with count different than 0 for the MSY genes.</p>
<p>Note that from the 34 MSY genes, only 33 are present in our datasets.</p>
<div id="version-with-msy-genes-only" class="section level4" number="3.1.2.1">
<h4><span class="header-section-number">3.1.2.1</span> Version with MSY genes only</h4>
<pre class="r"><code># Also include text for XIE genes.
MSY.genes &lt;- readRDS(&quot;MSYgenes.rds&quot;)
XIE.genes &lt;- readRDS(&quot;XIEgenes.rds&quot;)
length(MSY.genes)
[1] 34
length(XIE.genes)
[1] 66
mask &lt;- rownames(seB.filt.testing) %in% MSY.genes
sum(mask)
[1] 8

counts.MSY &lt;- seB.filt.testing[mask,]
seB.filt.testing$Sex &lt;- as.factor(seB.filt.testing$Sex)
assay.counts.MSY &lt;- assay(counts.MSY)
dataframe.counts.MSY &lt;- as.data.frame(t(assay.counts.MSY))

dataframe.counts.MSY$Sex &lt;- seB.filt.testing$Sex

data.imputed.MSY &lt;- kNN(dataframe.counts.MSY, k=3, variable=&quot;Sex&quot;)
seB.filt.testing$Sex &lt;- data.imputed.MSY$Sex</code></pre>
<pre class="r"><code>data.imputed.MSY$Sum &lt;- rowSums(data.imputed.MSY[,1:8])</code></pre>
</div>
<div id="version-with-xie-genes-only" class="section level4" number="3.1.2.2">
<h4><span class="header-section-number">3.1.2.2</span> Version with XIE genes only</h4>
<pre class="r"><code># Also include text for XIE genes.
#MSY.genes &lt;- readRDS(&quot;MSYgenes.rds&quot;)
#XIE.genes &lt;- readRDS(&quot;XIEgenes.rds&quot;)
#length(MSY.genes)
#length(XIE.genes)
#mask &lt;- rownames(seB.filt.testing) %in% XIE.genes
#sum(mask)

#counts.XIE &lt;- seB.filt.testing[mask,]
#seB.filt.testing$Sex &lt;- as.factor(seB.filt.testing$Sex)
#assay.counts.XIE &lt;- assay(counts.XIE)
#dataframe.counts.XIE &lt;- as.data.frame(t(assay.counts.XIE))

#dataframe.counts.XIE$Sex &lt;- seB.filt.testing$Sex

#data.imputed.XIE &lt;- kNN(dataframe.counts.XIE, k=3, variable=&quot;Sex&quot;)
#seB.filt.testing$Sex &lt;- data.imputed.XIE$Sex</code></pre>
<pre class="r"><code>#dataframe.counts.XIE$Sum &lt;- rowSums(dataframe.counts.XIE[,1:39])</code></pre>
<p>Save <code>DGEList</code> and <code>SummarizedExperiment</code> objects including filtered data,
normalization factors and normalized log CPM units of expression.</p>
<pre class="r"><code>saveRDS(dgeR.filt, file=file.path(&quot;_processed_data&quot;, &quot;dgeR.filt.rds&quot;))
saveRDS(seR.filt, file=file.path(&quot;_processed_data&quot;, &quot;seR.filt.rds&quot;))

saveRDS(dgeB.filt, file=file.path(&quot;_processed_data&quot;, &quot;dgeB.filt.rds&quot;))
saveRDS(seB.filt, file=file.path(&quot;_processed_data&quot;, &quot;seB.filt.rds&quot;))

saveRDS(dgeR.filt.training,
        file=file.path(&quot;_processed_data&quot;, &quot;dgeR.filt.training.rds&quot;))
saveRDS(seR.filt.training,
        file=file.path(&quot;_processed_data&quot;, &quot;seR.filt.training.rds&quot;))
saveRDS(dgeB.filt.testing,
        file=file.path(&quot;_processed_data&quot;, &quot;dgeB.filt.testing.rds&quot;))
saveRDS(seB.filt.testing,
        file=file.path(&quot;_processed_data&quot;, &quot;seB.filt.testing.rds&quot;))</code></pre>
</div>
</div>
</div>
</div>
<div id="dispersion-estimates-and-hypervariable-genes" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Dispersion estimates and hypervariable genes</h1>
<p>We calculate dispersion estimates and biological coefficients of variation
<span class="citation">(<a href="#ref-mccarthy2012differential">McCarthy, Chen, and Smyth 2012</a>)</span> and use them to identify hypervariable genes for
different models of increasing complexity that will help us to select the most
sensible one for the differential expression analysis between
Cancer-affected and unafected individuals.</p>
<p>The four considered models are: (1) a model with the intercept term only,
(2) a model with <code>Tumor</code> as main explanatory variable, (3) a model
with <code>Tumor</code> as main explanatory variable and <code>Sex</code> as
covariate, and (4) a model with <code>Tumor</code> as main explanatory variable, and <code>Sex</code> and surrogate variables (SVs), calculated with SVA
<span class="citation">(<a href="#ref-leek2007capturing">Leek and Storey 2007</a>)</span> as covariates. The estimation of SVs will be done using
only the expression values of the top-10% most variable genes.</p>
<div id="training-dataset" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Training dataset</h2>
<pre class="r"><code>library(sva)

mod &lt;- model.matrix(~ 1, colData(seR.filt.training))
dgeR.filt.training.interceptonly &lt;- estimateDisp(dgeR.filt.training, design=mod,
                                                 robust=TRUE)
summary(dgeR.filt.training.interceptonly$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  4.443   4.948   4.948   4.871   4.948   4.948 

mod &lt;- model.matrix(~ Tumor, colData(seR.filt.training))
dgeR.filt.training.PE &lt;- estimateDisp(dgeR.filt.training, design=mod,
                                      robust=TRUE)
summary(dgeR.filt.training.PE$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  4.453   4.969   4.969   4.890   4.969   4.969 

mod &lt;- model.matrix(~ Tumor + Sex, colData(seR.filt.training))
dgeR.filt.training.PEGA &lt;- estimateDisp(dgeR.filt.training, design=mod,
                                        robust=TRUE)
summary(dgeR.filt.training.PEGA$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  4.460   4.987   4.987   4.906   4.987   4.987 

mod0 &lt;- model.matrix(~ Sex, colData(seR.filt.training))
IQRs &lt;- apply(assays(seR.filt.training)$logCPM, 1, IQR)
mask &lt;- IQRs &gt; quantile(IQRs, prob=0.9)
sv &lt;- sva(assays(seR.filt.training)$logCPM[mask, ],
          mod=mod, mod0=mod0)
Number of significant surrogate variables is:  3 
Iteration (out of 5 ):1  2  3  4  5  
mod &lt;- cbind(mod, sv$sv)
dgeR.filt.training.PEGASVs &lt;- estimateDisp(dgeR.filt.training,
                                           design=mod, robust=TRUE)
summary(dgeR.filt.training.PEGASVs$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.3698 10.9890 10.9890 10.2764 10.9890 10.9890 </code></pre>
<p>Figure <a href="#fig:Mtrainingavglogcpmxvbcv">2</a> shows, for each considered model, the
biological coefficient of variation (BCV) as function of the average expression,
in log CPM units.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:Mtrainingavglogcpmxvbcv"></span>
<img src="03_qa_cn_files/figure-html/Mtrainingavglogcpmxvbcv-1.png" alt="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units." width="100%"  class="widefigure" />
<p class="caption">
Figure 2: <span class="caption-title">Biological coefficient of variation (BCV)</span><br>BCV values as function of the average gene expression in log-CPM units.
</p>
</div>
<p>Table <a href="#tab:hypervargenesMtrainingcn">1</a> below shows the top-20 hypervariable genes
with lowest prior degrees of freedom for the last model.</p>
<div class='horizontal-scroll'><table class="table table-striped table-hover table-responsive table" style="margin-left: auto; margin-right: auto; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:hypervargenesMtrainingcn">Table 1: </span>Top-20 hypervariable genes for the model Tumor + Sex + SVs.
</caption>
<thead>
<tr>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
prior.df
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
BCV
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Gene
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Description
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
0.3698417
</td>
<td style="text-align:right;">
1.3469176
</td>
<td style="text-align:left;">
RPL23AP25
</td>
<td style="text-align:left;">
ribosomal protein L23a pseudogene 25
</td>
</tr>
<tr>
<td style="text-align:right;">
0.3698417
</td>
<td style="text-align:right;">
1.2093621
</td>
<td style="text-align:left;">
RPL23AP7
</td>
<td style="text-align:left;">
ribosomal protein L23a pseudogene 7
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4558978
</td>
<td style="text-align:right;">
0.8648419
</td>
<td style="text-align:left;">
RMRP
</td>
<td style="text-align:left;">
RNA component of mitochondrial RNA processing endoribonuclease
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4558978
</td>
<td style="text-align:right;">
0.9898510
</td>
<td style="text-align:left;">
HBG1
</td>
<td style="text-align:left;">
hemoglobin subunit gamma 1
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4558978
</td>
<td style="text-align:right;">
1.0706268
</td>
<td style="text-align:left;">
HBG2
</td>
<td style="text-align:left;">
hemoglobin subunit gamma 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4558978
</td>
<td style="text-align:right;">
0.9374988
</td>
<td style="text-align:left;">
CENPBD2P
</td>
<td style="text-align:left;">
CENPB DNA-binding domains containing 2, pseudogene
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4745407
</td>
<td style="text-align:right;">
1.1543784
</td>
<td style="text-align:left;">
RNU5A-1
</td>
<td style="text-align:left;">
RNA, U5A small nuclear 1
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4912080
</td>
<td style="text-align:right;">
0.7946923
</td>
<td style="text-align:left;">
RNU11
</td>
<td style="text-align:left;">
RNA, U11 small nuclear
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4912080
</td>
<td style="text-align:right;">
0.8668384
</td>
<td style="text-align:left;">
TRNG
</td>
<td style="text-align:left;">
tRNA-Gly
</td>
</tr>
<tr>
<td style="text-align:right;">
0.4958599
</td>
<td style="text-align:right;">
0.8424359
</td>
<td style="text-align:left;">
SNORD3C
</td>
<td style="text-align:left;">
small nucleolar RNA, C/D box 3C
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5179521
</td>
<td style="text-align:right;">
1.3980551
</td>
<td style="text-align:left;">
RNF182
</td>
<td style="text-align:left;">
ring finger protein 182
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5179521
</td>
<td style="text-align:right;">
1.0657600
</td>
<td style="text-align:left;">
RNU5B-1
</td>
<td style="text-align:left;">
RNA, U5B small nuclear 1
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5179521
</td>
<td style="text-align:right;">
0.9582668
</td>
<td style="text-align:left;">
SNORD3A
</td>
<td style="text-align:left;">
small nucleolar RNA, C/D box 3A
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5244248
</td>
<td style="text-align:right;">
0.8066638
</td>
<td style="text-align:left;">
MAGI2-AS3
</td>
<td style="text-align:left;">
MAGI2 antisense RNA 3
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5244248
</td>
<td style="text-align:right;">
0.8584075
</td>
<td style="text-align:left;">
DEFA1B
</td>
<td style="text-align:left;">
defensin alpha 1B
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5244248
</td>
<td style="text-align:right;">
0.9654232
</td>
<td style="text-align:left;">
TRNL2
</td>
<td style="text-align:left;">
tRNA-Leu
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5380500
</td>
<td style="text-align:right;">
0.7573270
</td>
<td style="text-align:left;">
ALB
</td>
<td style="text-align:left;">
albumin
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5380500
</td>
<td style="text-align:right;">
0.7265875
</td>
<td style="text-align:left;">
FHDC1
</td>
<td style="text-align:left;">
FH2 domain containing 1
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5380500
</td>
<td style="text-align:right;">
0.7070285
</td>
<td style="text-align:left;">
ALAS2
</td>
<td style="text-align:left;">
5’-aminolevulinate synthase 2
</td>
</tr>
<tr>
<td style="text-align:right;">
0.5380500
</td>
<td style="text-align:right;">
0.8982991
</td>
<td style="text-align:left;">
TRNT
</td>
<td style="text-align:left;">
tRNA-Thr
</td>
</tr>
</tbody>
</table></div>
</div>
<div id="testing-data" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Testing data</h2>
<pre class="r"><code>mod &lt;- model.matrix(~ 1, colData(seB.filt.testing))
dgeB.filt.testing.interceptonly &lt;- estimateDisp(dgeB.filt.testing, design=mod,
                                                robust=TRUE)
summary(dgeB.filt.testing.interceptonly$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  11.80   11.80   11.80   11.99   12.31   12.47 
mod &lt;- model.matrix(~ Tumor, colData(seB.filt.testing))
dgeB.filt.testing.PE &lt;- estimateDisp(dgeB.filt.testing, design=mod,
                                     robust=TRUE)
summary(dgeB.filt.testing.PE$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  13.67   13.67   13.67   13.89   14.30   14.41 
mod &lt;- model.matrix(~ Tumor + Sex, colData(seB.filt.testing))
dgeB.filt.testing.PEGA &lt;- estimateDisp(dgeB.filt.testing, design=mod,
                                       robust=TRUE)
summary(dgeB.filt.testing.PEGA$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  14.38   14.38   14.38   14.63   15.13   15.18 
mod0 &lt;- model.matrix(~ Sex, colData(seB.filt.testing))
IQRs &lt;- apply(assays(seB.filt.testing)$logCPM, 1, IQR)
mask &lt;- IQRs &gt; quantile(IQRs, prob=0.9)
sv &lt;- sva(assays(seB.filt.testing)$logCPM[mask, ],
          mod=mod, mod0=mod0)
Number of significant surrogate variables is:  2 
Iteration (out of 5 ):1  2  3  4  5  
mod &lt;- cbind(mod, sv$sv)
dgeB.filt.testing.PEGASVs &lt;- estimateDisp(dgeB.filt.testing,
                                          design=mod, robust=TRUE)
summary(dgeB.filt.testing.PEGASVs$prior.df)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  4.579  31.138  31.138  30.376  31.138  31.138 </code></pre>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:Mtestingavglogcpmxvbcv"></span>
<img src="03_qa_cn_files/figure-html/Mtestingavglogcpmxvbcv-1.png" alt="Biological coefficient of variation (BCV). BCV values as function of the average gene expression in log-CPM units." width="100%"  class="widefigure" />
<p class="caption">
Figure 3: <span class="caption-title">Biological coefficient of variation (BCV)</span><br>BCV values as function of the average gene expression in log-CPM units.
</p>
</div>
<div class='horizontal-scroll'><table class="table table-striped table-hover table-responsive table" style="margin-left: auto; margin-right: auto; margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:hypervargenesDsubset">Table 2: </span>Top-20 hypervariable genes for the model Tumor + Sex + SVs.
</caption>
<thead>
<tr>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
prior.df
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
BCV
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Gene
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Description
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
4.579091
</td>
<td style="text-align:right;">
1.615303
</td>
<td style="text-align:left;">
PRG4
</td>
<td style="text-align:left;">
proteoglycan 4
</td>
</tr>
<tr>
<td style="text-align:right;">
4.579091
</td>
<td style="text-align:right;">
1.143039
</td>
<td style="text-align:left;">
ALB
</td>
<td style="text-align:left;">
albumin
</td>
</tr>
<tr>
<td style="text-align:right;">
4.579091
</td>
<td style="text-align:right;">
1.203413
</td>
<td style="text-align:left;">
PPBP
</td>
<td style="text-align:left;">
pro-platelet basic protein
</td>
</tr>
<tr>
<td style="text-align:right;">
4.579091
</td>
<td style="text-align:right;">
1.286772
</td>
<td style="text-align:left;">
MUC5B
</td>
<td style="text-align:left;">
mucin 5B, oligomeric mucus/gel-forming
</td>
</tr>
<tr>
<td style="text-align:right;">
4.777921
</td>
<td style="text-align:right;">
1.285881
</td>
<td style="text-align:left;">
SNORD17
</td>
<td style="text-align:left;">
small nucleolar RNA, C/D box 17
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.128590
</td>
<td style="text-align:left;">
CFH
</td>
<td style="text-align:left;">
complement factor H
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.341406
</td>
<td style="text-align:left;">
CRP
</td>
<td style="text-align:left;">
C-reactive protein
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.171850
</td>
<td style="text-align:left;">
F5
</td>
<td style="text-align:left;">
coagulation factor V
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.347031
</td>
<td style="text-align:left;">
CPS1
</td>
<td style="text-align:left;">
carbamoyl-phosphate synthase 1
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.054619
</td>
<td style="text-align:left;">
APOB
</td>
<td style="text-align:left;">
apolipoprotein B
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.313924
</td>
<td style="text-align:left;">
CP
</td>
<td style="text-align:left;">
ceruloplasmin
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.350498
</td>
<td style="text-align:left;">
CLDN1
</td>
<td style="text-align:left;">
claudin 1
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.212169
</td>
<td style="text-align:left;">
PF4V1
</td>
<td style="text-align:left;">
platelet factor 4 variant 1
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.154580
</td>
<td style="text-align:left;">
FGB
</td>
<td style="text-align:left;">
fibrinogen beta chain
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.184251
</td>
<td style="text-align:left;">
TMEM271
</td>
<td style="text-align:left;">
transmembrane protein 271
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.245340
</td>
<td style="text-align:left;">
HMX1
</td>
<td style="text-align:left;">
H6 family homeobox 1
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.232104
</td>
<td style="text-align:left;">
FGG
</td>
<td style="text-align:left;">
fibrinogen gamma chain
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.295402
</td>
<td style="text-align:left;">
TAF11L4
</td>
<td style="text-align:left;">
TATA-box binding protein associated factor 11 like 4
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.276243
</td>
<td style="text-align:left;">
ORM1
</td>
<td style="text-align:left;">
orosomucoid 1
</td>
</tr>
<tr>
<td style="text-align:right;">
5.895835
</td>
<td style="text-align:right;">
1.243108
</td>
<td style="text-align:left;">
ITIH2
</td>
<td style="text-align:left;">
inter-alpha-trypsin inhibitor heavy chain 2
</td>
</tr>
</tbody>
</table></div>
</div>
</div>
<div id="session-information" class="section level1" number="5">
<h1><span class="header-section-number">5</span> Session information</h1>
<pre class="r"><code>sessionInfo()
R version 4.3.3 (2024-02-29)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.5.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Europe/Madrid
tzcode source: internal

attached base packages:
[1] stats4    grid      stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] plotrix_3.8-4               sva_3.50.0                 
 [3] BiocParallel_1.36.0         genefilter_1.84.0          
 [5] mgcv_1.9-1                  nlme_3.1-164               
 [7] edgeR_4.0.16                limma_3.58.1               
 [9] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[11] GenomicRanges_1.54.1        GenomeInfoDb_1.38.8        
[13] IRanges_2.36.0              S4Vectors_0.40.2           
[15] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[17] matrixStats_1.3.0           VIM_6.2.2                  
[19] colorspace_2.1-0            kableExtra_1.4.0           
[21] knitr_1.45                  BiocStyle_2.30.0           

loaded via a namespace (and not attached):
 [1] DBI_1.2.2               bitops_1.0-7            rlang_1.1.3            
 [4] magrittr_2.0.3          e1071_1.7-14            compiler_4.3.3         
 [7] RSQLite_2.3.6           png_0.1-8               systemfonts_1.0.6      
[10] vctrs_0.6.5             stringr_1.5.1           crayon_1.5.2           
[13] fastmap_1.1.1           XVector_0.42.0          rmarkdown_2.26         
[16] bit_4.0.5               xfun_0.43               zlibbioc_1.48.2        
[19] cachem_1.0.8            jsonlite_1.8.8          blob_1.2.4             
[22] highr_0.10              DelayedArray_0.28.0     parallel_4.3.3         
[25] R6_2.5.1                bslib_0.7.0             stringi_1.8.3          
[28] vcd_1.4-12              ranger_0.16.0           car_3.1-2              
[31] boot_1.3-30             lmtest_0.9-40           jquerylib_0.1.4        
[34] Rcpp_1.0.12             bookdown_0.39           zoo_1.8-12             
[37] Matrix_1.6-5            splines_4.3.3           nnet_7.3-19            
[40] rstudioapi_0.16.0       abind_1.4-5             yaml_2.3.8             
[43] codetools_0.2-20        lattice_0.22-6          KEGGREST_1.42.0        
[46] evaluate_0.23           survival_3.6-4          proxy_0.4-27           
[49] xml2_1.3.6              Biostrings_2.70.3       BiocManager_1.30.22    
[52] carData_3.0-5           sp_2.1-3                RCurl_1.98-1.14        
[55] munsell_0.5.1           scales_1.3.0            laeken_0.5.3           
[58] xtable_1.8-4            class_7.3-22            glue_1.7.0             
[61] tools_4.3.3             robustbase_0.99-2       data.table_1.15.4      
[64] annotate_1.80.0         locfit_1.5-9.9          XML_3.99-0.16.1        
[67] AnnotationDbi_1.64.1    GenomeInfoDbData_1.2.11 cli_3.6.2              
[70] S4Arrays_1.2.1          viridisLite_0.4.2       svglite_2.1.3          
[73] DEoptimR_1.1-3          sass_0.4.9              digest_0.6.35          
[76] SparseArray_1.2.4       memoise_2.0.1           htmltools_0.5.8.1      
[79] lifecycle_1.0.4         httr_1.4.7              statmod_1.5.0          
[82] bit64_4.0.5             MASS_7.3-60.0.1        </code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Block2022" class="csl-entry">
Block, Timothy, Daniel Zezulinski, David Kaplan, Jingqiao Lu, Samantha Zanine, Tingting Zhan, Cataldo Doria, and Aejaz Sayeed. 2022. <span>“Circulating Messenger RNA Variants as a Potential Biomarker for Surveillance of Hepatocellular Carcinoma.”</span> <em>Frontiers in Oncology</em> 12. <a href="https://doi.org/10.3389/fonc.2022.963641">https://doi.org/10.3389/fonc.2022.963641</a>.
</div>
<div id="ref-leek2007capturing" class="csl-entry">
Leek, Jeffrey T, and John D Storey. 2007. <span>“Capturing Heterogeneity in Gene Expression Studies by Surrogate Variable Analysis.”</span> <em>PLOS Genetics</em> 3 (9): e161.
</div>
<div id="ref-mccarthy2012differential" class="csl-entry">
McCarthy, Davis J, Yunshun Chen, and Gordon K Smyth. 2012. <span>“Differential Expression Analysis of Multifactor RNA-Seq Experiments with Respect to Biological Variation.”</span> <em>Nucleic Acids Research</em> 40 (10): 4288–97.
</div>
<div id="ref-Roskams2022" class="csl-entry">
Roskams-Hieter, Breeshey, Hyun Ji Kim, Pavana Anur, Josiah T. Wagner, Rowan Callahan, Elias Spiliotopoulos, Charles Ward Kirschbaum, et al. 2022. <span>“Plasma Cell-Free RNA Profiling Distinguishes Cancers from Pre-Malignant Conditions in Solid and Hematologic Malignancies.”</span> <em>Npj Precision Oncology</em> 6. <a href="https://doi.org/10.1038/s41698-022-00270-y">https://doi.org/10.1038/s41698-022-00270-y</a>.
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": {
      styles: {
        ".MathJax_Display": {
           "text-align": "center",
           padding: "0px 150px 0px 65px",
           margin: "0px 0px 0.5em"
        },
        "@media screen and (max-width: 991px)": {
            ".MathJax_Display": {
               "text-align": "center",
               padding: "0 0 0 0"
            }
         }
      }
    }
  });
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<script type="text/javascript">
$(document).ready(function ()  {
  
  // Map "enter" keypress to the same action as a cursor click
  function navigateLink(e) {
    if (e.key === "Enter") {
      $(this).trigger("click");
    }
  }

  var toc_items = document.querySelectorAll(".tocify-item");
  for (var i = 0; i < toc_items.length; i++) {
    // The link role tells screen readers this is for navigation
    toc_items.item(i).setAttribute("role", "link");
    // tabindex = 0 allows selection via keyboard tab presses
    toc_items.item(i).setAttribute("tabindex", "0");
    // Listen for "Enter" keypress when item is selected
    toc_items.item(i).addEventListener("keydown", navigateLink);
  }
});
</script>

</body>
</html>
